---
title: "R Notebook"
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: console
---

## Load libraries
```{r warning=FALSE, message=FALSE}
library(tidyverse)
library(scran)
library(scater)
library(gprofiler2)
library(gridExtra)
library(ComplexHeatmap)
library(circlize)
library(edgeR)
library(readxl)
library(ggrepel)
library(tftargets)
library(network)
library(ggnet)
library(parallel)
library(BiocParallel)
library(ggsci)
library(airr)
library(numbat)
library(scales)

ncores <- 10
mcparam <- MulticoreParam(workers=ncores)
register(mcparam)

```

## Define variables
```{r}
opt <- list()
#opt$deres <- "data/TPLL_deres_dmso.csv"
#opt$loop2020 <- "data/de_res.csv"
opt$pathinfo <- "misc/pathways_2.xlsx"
opt$scebulk <- "data/submission/RNA_complete.RDS"
opt$plot <- "plots/Fig3/"


## Set theme
lgd <-  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5), 
        text = element_text(size = 15),
        panel.background = element_rect(fill = "transparent",colour = NA),
        plot.background = element_rect(fill = "transparent",colour = NA), 
        legend.background = element_rect(fill='transparent',colour = NA), 
        strip.background = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

```

## Load data
```{r}
## SCE
sce <- readRDS(opt$scebulk)
sce

# ##
# sce.sc <- readRDS(opt$scesc)
# sce.sc <- sce.sc[[1]]
# 
# ## single-cell cell types
# types <- read.csv(opt$types)
# 
# ## Load tcr data
# immTab <- read_rearrangement(opt$immtab)
# 
# ## Load differentially expressed genes
# de.res <- read.csv(opt$deres)

## Load pathway annotation
path.info <- read_excel(opt$pathinfo)
```

## Load transcription factors
```{r}
# ntop <- 1000
# maxtargets <- 5000
# 
# ## Get TF target numbers
# motif.df <- lapply(names(Marbach2016), function(x) {
#   tf.list <- Marbach2016[[x]]
#   #tf.list <- TRRUST[[x]]
#   
#   data.frame(motif_name = x, n_targets = length(tf.list))
#   
# }) %>% bind_rows()
# 
# ## Get TF targets
# target.df <- lapply(names(Marbach2016), function(x) {
#   tf.list <- Marbach2016[[x]]
#   
#   data.frame(motif_name = x, targets = tf.list)
#   
# }) %>% bind_rows()

```

## Analysis
#### Perform differential gene expression testing
```{r}
# ## Remove Ig genes
# igh <- grep("^IGH", rowData(sce)$Symbol)
# igl <- grep("^IGL", rowData(sce)$Symbol)
# igk <- grep("^IGK", rowData(sce)$Symbol)
# igk <- grep("^IGK", rowData(sce)$Symbol)
# tra <- grep("^TRA", rowData(sce)$Symbol)
# trb <- grep("^TRB", rowData(sce)$Symbol)
# trg <- grep("^TRG", rowData(sce)$Symbol)
# trd <- grep("^TRD", rowData(sce)$Symbol)
#     
# igGenes <- rownames(sce)[c(igh, igl, igk, tra, trb, trg, trd)]
# #message(igGenes)
# message("Removing ", length(igGenes), " genes")
# remGenes <- rownames(sce) %in% igGenes
# 
# hvgenes <- rownames(sce)[!remGenes]


testTreat <- unique(sce$Treatment)

de.res <- mclapply("DMSO", mc.cores = ncores, 
                 function(z) {
                   
  message(paste0("Fitting model for ", z))
                   
  compDrug <- testTreat[testTreat != z] # instead of subtypes we use other treatments here
                   
  mclapply(compDrug, mc.cores = ncores, 
         function(x) {
    message(paste0("Drug ", x))
    #sce.x <- sce.x[, sce.x$celltype_broad == x & sce.x$Treatment %in% c("Untreated", compDrug)]
    #sce.x <- sce.x[, sce.x$patientID %in% pairedPat]
    # summed <- summarizeAssayByGroup(sce.x, ids = colData(sce.x)[,c("patientID", "sampleID", "Treatment")],
    #                                 statistics = "sum", assay.type = "counts") # get raw counts and sum them up
    summed <- sce#[hvgenes, ]
    colData(summed)$Treatment <- factor(colData(summed)$Treatment, levels = c(z, compDrug))
    y <- DGEList(counts(summed), samples = colData(summed))
    # discarded <- summed$ncells < 10
    # y <- y[,!discarded]
    # message(paste0("Discarding ", sum(discarded), " samples"))
    design <- model.matrix(~0 + factor(patientID) + Treatment, y$samples) # set coefficient (name or matrix column) - fig. 7 f1000
    keep <- filterByExpr(y, design = design)
    #keep <- filterByExpr(y, group = design)
    y <- y[keep,]
    summary(keep)
    message(paste0("Keeping ", sum(keep), " genes"))
    y <- calcNormFactors(y)
    y <- estimateDisp(y, design) # gene-wise dispersion (variance on top of poisson distribution)
    plotBCV(y)
    fit <- glmQLFit(y, design, robust = TRUE) # QL way of fitting the model
    plotQLDisp(fit)
    res <- glmQLFTest(fit, coef = paste0("Treatment", x))
    summary(decideTests(res))
    resTab <- res$table
    resTab$gene <- rownames(resTab)
    resTab <- resTab %>% mutate(p.adj = p.adjust(PValue, method = "BH")) #%>%
      #mutate(p.adj = ifelse(p.adj < 1e-12, 1e-12, p.adj))
    resTab$Treatment <- x
    resTab$Control <- z
    resTab %>% arrange(desc(logFC))
  }) %>% bind_rows()
}) %>% bind_rows()

de.res <- de.res %>%
  left_join(select(data.frame(rowData(sce)), ID, Symbol), c("gene" = "ID"))

de.res %>% filter(Treatment == "Selinexor", Symbol %in% c("BCL2", "TP53"))
#write.csv(de.res, opt$deres)

```


#### Show the number of DEG
```{r}
sigTab <- de.res %>%
  group_by(Treatment) %>%
  filter(p.adj < 0.1) %>%
  filter(logFC > 0.25 | logFC < -0.25) %>%
  mutate(dir = ifelse(logFC > 0, "Up", "Down")) %>%
  dplyr::count(dir) %>%
  dplyr::rename(Direction = dir) %>%
  ungroup()

drugOrder <- sigTab %>%
  group_by(Treatment) %>%
  summarise(allDEG = sum(n)) %>%
  arrange(desc(allDEG)) %>%
  pull(Treatment)

sigTab <- sigTab %>%
  mutate(Treatment = factor(Treatment, levels = drugOrder), 
         n = ifelse(Direction == "Down", n * -1, n))

p <- sigTab %>%
  ggplot(aes(x = Treatment, y = n, fill = Direction)) +
  geom_bar(stat = "identity", colour = "black") +
  geom_text(aes(label = n), data = sigTab[sigTab$Direction == "Up", ], size = 5, nudge_y = 200) +
  geom_text(aes(label = n), data = sigTab[sigTab$Direction == "Down", ], size = 5, nudge_y = -200) +
  xlab("") + ylab("Number of Genes") +
  ggtitle("Significantly Differentially Expressed Genes") +
  lgd + scale_fill_manual(values = c("Up" = "#F44336", "Down" = "#1976D2")) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1), 
        plot.title = element_text(size = 21, vjust = 1), 
        axis.text = element_text(size = 17.5), 
        text = element_text(size = 22.5), 
        panel.border = element_rect(linewidth = 1))
p
ggsave(plot = p, filename = paste0(opt$plot, "n_deg.png"), 
       height = 17.5, width = 22, units = "cm")

```


#### Construct a network of differentially expressed genes
```{r fig.height= 15, fig.width=15}
# Assign nodes (treatments) and edges (5% FDR DEGs)
plotMat <- de.res %>% filter(p.adj < 0.1) %>%
  filter(logFC > 0.25 | logFC < -0.25)
allTreat <- unique(plotMat$Treatment)
target <- data.frame(Treatment = allTreat,
                     target = c("Autophagyi", "IAPi", "HDACi", "mTORi", "ITKi", "TLR8a", "MDM2i", 
                                "IMID", "JAKi", "XPOi", "BCL2i"))

opt$sub <- FALSE

if(opt$sub == TRUE) {
  plotMat <- de.res %>% filter(p.adj < 0.1, !Treatment %in% c("Pomalidomide", "Dacinostat")
                             ) %>%
  filter(logFC > 0.25 | logFC < -0.25) #%>%
    #filter(logFC < -0.25)
} else {
  plotMat <- de.res %>% filter(p.adj < 0.1) %>%
    filter(logFC > 0.25 | logFC < -0.25) #%>%
    #filter(logFC > 0.25)
}


allTreat <- unique(plotMat$Treatment)

nodes <- data.frame(Treatment = allTreat)
edges <- plotMat %>% select(Treatment, Symbol) %>%
  mutate(Treatment = as.character(Treatment))

# Create network object
set.seed(100)
net <- network(edges, directed = TRUE)
netTab <- ggnet2(net)$data |> 
  mutate(Type = ifelse(label %in% allTreat, "Treatment", "Symbol"),
         Treatment = ifelse(Type == "Treatment", label, NA)) |> 
  select(-alpha,-color,-shape,-size) |> 
  mutate(x = as.numeric(x), y = as.numeric(y))
  
# Define edges based on direction and padj of DEGs
from <- edges %>%
  left_join(netTab, by = c("Treatment" = "label")) %>% dplyr::select(Treatment, x, y) %>%
  dplyr::rename(x_start = x, y_start = y) |> distinct()

to <- edges %>%
  left_join(netTab, by = c("Symbol" = "label")) %>% dplyr::select(Symbol,x,y) %>%
  dplyr::rename(x_end = x, y_end = y) |> distinct()

edges <- edges %>% left_join(from) %>% left_join(to) %>%
  left_join(plotMat %>% select(Symbol,Treatment,logFC,p.adj)) %>%
  distinct() %>%
  mutate(Direction = ifelse(logFC < 0,"Down","Up"),
         alpha = -log10(p.adj)
         #alpha = rescale(-log10(p.adj#squish(p.adj, c(1.812407e-209,1))))
         )

## Joining with `by = join_by(treatment)`
## Joining with `by = join_by(symbol)`
## Joining with `by = join_by(treatment, symbol)`

# Create auxiliary table for annotation

n_DEG <- plotMat %>% #filter(!Treatment %in% c("Pomalidomide", "BafilomycinA1", "Dacinostat")) %>%
  group_by(Treatment) %>%
  filter(p.adj < 0.1) %>%
  filter(logFC > 0.25 | logFC < -0.25) %>% 
  dplyr::count() %>% ungroup()

netTabAnno <- netTab |> 
  left_join(target) |> 
  left_join(n_DEG) |> 
  mutate(Treatment = factor(Treatment, levels = allTreat))

p <- ggplot()  +
  geom_curve(data = edges,
               aes(x = x_start, xend = x_end, y = y_start, yend = y_end, col = Direction, alpha = p.adj),
               linewidth = 0.1, curvature = 0.05) +
  geom_point(data = netTabAnno[netTabAnno$Type == "Symbol",],
             aes(x = x, y = y),
             size = 0.5, shape = 21, color = "grey60", fill = "grey95", stroke = 0.75, alpha = 0.6) +
  geom_point(data = netTabAnno[netTabAnno$Type == "Treatment",],
             aes(x = x, y = y, fill = Treatment, size = n), shape = 21, stroke = 0.75) +
  # scale_fill_manual(values = col_TX) +
  scale_size_continuous(name = "Number of DEGs",
                        breaks = seq(0, max(n_DEG$n), by = 1000), limits = c(0, max(n_DEG$n)), range = c(0.5,7.5)) +
  scale_color_manual(values = c("Up" = "#F44336", "Down" = "#1976D2")) +
  scale_alpha_continuous() +
  geom_label_repel(data = netTabAnno |> filter(Type == "Treatment"),
                  aes(x = x, y = y, label = Treatment, fill = Treatment, point.size = rescale(n, to = c(0.5,7.5))),
                  max.overlaps = Inf, vjust = "bottom", nudge_y = -0.001, fontface = "bold", alpha = 0.7, #direction = "y", 
                  show.legend = FALSE, size = 5)  +
  guides(color = guide_legend(ncol = 2, override.aes = list(linewidth=2.5)),
         size = guide_legend(ncol = 2),
         alpha = guide_legend(override.aes = list(linewidth=2.5)),
         fill = guide_legend(override.aes = list(size=3.5))) +
  xlab("") + ylab("") + ggtitle("Network of Differentially Expressed Genes") +
  lgd +
  theme(axis.ticks = element_blank(), 
        axis.text = element_blank(), 
        panel.border = element_rect(linewidth = 1, colour = NA, fill = "transparent")) +
  scale_fill_manual(values = c("Venetoclax" = "grey70", "Birinapant" = "#AD1457", "Nutlin-3a" = "#FFA000", 
                                "Ibrutinib" = "#283593", "Everolimus" = "#43A047", "Selinexor" = "#FFD45F",
                               "BafilomycinA1" = "#F8BBD0", "Motolimod" = "#AB47BC", "Dacinostat" = "#F44336", 
                                "Dacinostat" = "#64B5F6", "Pomalidomide" = "#80DEEA", "Ruxolitinib" = "#4DB6AC"))

p
if(opt$sub == TRUE) {
 ggsave(plot = p, filename = paste0(opt$plot, "deg_subnetwork.png"),
       height = 20, width = 25, units = "cm")
} else {
  ggsave(plot = p, filename = paste0(opt$plot, "deg_network.png"),
       height = 20, width = 25, units = "cm")
}


```

#### Volcano plots
```{r}

pList <- lapply(c("Birinapant", "Selinexor", "Nutlin-3a"), function(x) {
  
  if(x %in% c("Motolimod", "Everolimus", "Ibrutinib", "BaflimoycinA1")) {
    geneSelec <- path.info %>% filter(pathway %in% c("NFKB", "NFKB_Chemo", "NFKB_GF", "NFKB_TF", "NFKB_Apo", "TNF", "TLR")) %>% select(gene) %>% unlist() %>% as.vector()
  } else if(x %in% c("Birinapant")) {
    geneSelec <- path.info %>% filter(pathway %in% c("NFKB", "NFKB_Chemo", "NFKB_GF", "NFKB_TF", "NFKB_Apo", "TNF")) %>% select(gene) %>% unlist() %>% as.vector()
  } else if(x %in% c("Nutlin-3a", "Selinexor")) {
    geneSelec <- path.info %>% filter(pathway %in% c("TP53up")) %>% select(gene) %>% unlist() %>% as.vector()
    geneSelec <- c("TP53", geneSelec)

  } else if(x %in% c("Everolimus")) {
    geneSelec <- path.info %>% filter(pathway %in% c("NFKB", "NFKB_Chemo", "NFKB_GF", "NFKB_TF", "NFKB_Apo", "TNF", "TLR")) %>% select(gene) %>% unlist() %>% as.vector()
  }
  
  volTab <- de.res %>% filter(Treatment == x)
  #volTab <- de.res.2020 %>% filter(Treatment == x, celltype == "T-PLL") %>% dplyr::rename(Symbol = gene)
  
  ## Define limits  
  minLim <- -1 * max(abs(de.res$logFC))
  maxLim <- max(abs(de.res$logFC))
  thresh <- 0.25

  volTab$diffCol <- NA
  volTab[volTab$logFC > 0 & volTab$p.adj < 0.1 & volTab$logFC > thresh, "diffCol"] <- "sigUp"
  volTab[volTab$logFC < 0 & volTab$p.adj < 0.1 & volTab$logFC < -thresh, "diffCol"] <- "sigDown"
  volTab[volTab$p.adj >= 0.1, "diffCol"] <- "ns"
  volTab[volTab$logFC < thresh & volTab$logFC > -(thresh), "diffCol"] <- "ns"
  
  p <- volTab %>%
    mutate(logFC = ifelse(logFC > maxLim, maxLim, logFC)) %>%
    mutate(logFC = ifelse(logFC < minLim, minLim, logFC)) %>%
    #filter(gene %in% geneSelec) %>%
    filter(Treatment == x) %>%
    ggplot(aes(x = logFC, y = -log10(p.adj), fill = diffCol)) +
    geom_point(shape = 21, size = 2) + xlab("Log2FC") + ylab("-Log10(p.adj)") + ggtitle(paste0(x)) +
    #xlim(minLim, maxLim) +
    geom_hline(yintercept = -log10(0.1), linetype = "dashed") +
    geom_vline(xintercept = -thresh, linetype = "dashed") + geom_vline(xintercept = thresh, linetype = "dashed") +
    geom_label_repel(data = filter(volTab, Symbol %in% c(geneSelec), Treatment %in% x, p.adj < 0.1, 
                                 logFC > thresh | logFC < -(thresh)), aes(label = Symbol), fill = "white", 
                    nudge_y = 0.25, size = 4, max.overlaps = getOption("ggrepel.max.overlaps", default = 30), 
                     segment.linetype = 2, force = 20, alpha = 0.8) +
    scale_fill_manual(values = c("sigUp" = "#F44336", "sigDown" = "#1976D2", "ns" = "grey")) +
    theme_classic() +
    theme(plot.title = element_text(hjust = 0.5), 
          legend.position = "none", 
          text = element_text(size = 17.5), 
          axis.text = element_text(size = 15),
          panel.background = element_rect(fill = "transparent",colour = NA),
          plot.background = element_rect(fill = "transparent",colour = NA),
          legend.background = element_rect(fill='transparent'),
          strip.background = element_blank())
  
   ggsave(plot = p, filename = paste0(opt$plot, "vol_", x, ".png"),
          height = 15, width = 15, unit = "cm")
  p

})
pList 

geneSelec <- path.info %>% filter(pathway %in% c("NFKB", "NFKB_Chemo", "NFKB_GF", "NFKB_TF", "NFKB_Apo", "TNF", "TLR")) %>% select(gene) %>% unlist() %>% as.vector()

geneSub <- lapply(unique(de.res$Treatment), function(x) {
  de.res %>% 
    filter(Treatment == x) %>%
    filter(p.adj < 0.1) %>%
    filter(logFC > 0.25 | logFC < -0.25) %>%
    pull(Symbol)
  
}) %>% unlist() %>% unique()

geneSub <- intersect(geneSelec, geneSub)

plotMat.wide <- de.res %>% 
  mutate(logFC = ifelse(p.adj >= 0.1, 0, logFC)) %>%
  dplyr::select(Symbol, Treatment, logFC) %>% 
  filter(Symbol %in% geneSub) %>%
  dplyr::rename(log2FC = logFC) %>%
  pivot_wider(names_from = "Treatment", values_from = "log2FC") %>%
  column_to_rownames("Symbol")

geneOrder <- rownames(plotMat.wide)[hclust(dist(plotMat.wide), method = "ward.D2")$order]

p <- de.res %>% filter(#Treatment %in% c("Birinapant", "Motolimod"),
  Symbol %in% geneSub) %>%
  filter(p.adj < 0.1) %>%
  mutate(pSign = -log10(p.adj) * sign(logFC),
         p.adj = ifelse(-log10(p.adj) > 5, 5, -log10(p.adj)), 
         Symbol = factor(Symbol, levels = geneOrder)) %>%
  dplyr::rename(log2FC = logFC) %>%
  ggplot(aes(x = Symbol, y = Treatment)) +
  geom_point(aes(size = p.adj, fill = log2FC), shape = 21) +
  xlab("") + ylab("") + ggtitle("Effect on TNF-NFkB and TLR Target Genes") +
  lgd +
  scale_fill_gradient2(high = "#F44336", low = "#1976D2", name = "log2FC",
                       limits=c(-2, 2), oob=squish) +
  scale_size_continuous(name = "-Log10(p.adj)",
                        breaks = seq(0, 5, by = 2.5),
                        limits = c(0, 5), #range = c(0.5,7.5)
                        ) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        text = element_text(size = 12.5),
        panel.border = element_rect(linewidth = 1))
p
ggsave(plot = p, filename = paste0(opt$plot, "biri_moto_nfkb.png"),
       height = 9, width = 50, units = "cm")

# 
# plotMat.wide <- de.res %>% 
#   mutate(logFC = ifelse(p.adj >= 0.1, 0, logFC)) %>%
#   dplyr::select(Symbol, Treatment, logFC) %>% 
#   filter(Symbol %in% geneSub) %>%
#   dplyr::rename(log2FC = logFC) %>%
#   pivot_wider(names_from = "Treatment", values_from = "log2FC") %>%
#   column_to_rownames("Symbol")
# 
# #plotMat.wide <- scale(t(plotMat.wide))
# 
# pHeat <- Heatmap(t(plotMat.wide), 
#         show_row_names = TRUE,
#         cluster_rows = TRUE, border = TRUE, #column_names_gp = gpar(fontsize = 8),
#         #show_column_names = FALSE, 
#         cluster_columns = TRUE, column_names_gp = gpar(fontsize = 10),
#         clustering_method_rows = "ward.D2", #column_title = paste0(z),
#         clustering_method_columns = "ward.D2",
#         col=colorRamp2(c(-3, 0, 3), colors = c("#1976D2", "white", "#F44336")), name = "log2FC")
# pHeat
# # png(file=paste0(opt$plot, "nfkb_logFC_heatmap.png"),
# #     height = 10, width = 40, res = 600, units = "cm", bg = "transparent")
# # pHeat
# # draw(pHeat, background = "transparent")
# # dev.off()
#   
# pHeat

```

#### Heatmap
```{r}
# compDrug <- c("BafilomycinA1", "Selinexor", "Nutlin-3a")
# 
# diffMat.list <- lapply(compDrug, function(x) {
#   message(x)
#   geneSelec <- volTab <- de.res %>% filter(Treatment == x, p.adj < 0.1) %>%
#     filter(logFC > 0.25 | logFC < -0.25) %>% pull(gene)
#   
#   ## Get the count matrix
#   count.mat <- logcounts(sce[geneSelec, sce$Treatment %in% c("DMSO", x)])
#   
#   ## Compute the logFC per patient
#   diffMat <- lapply(unique(sce$patientID), function(x) {
#     ## Subset to the samples of this patient
#     matSub <- count.mat[, str_detect(string = colnames(count.mat), pattern = x)]
#     
#     ## Save gene names
#     matSub <- matSub %>% data.frame() %>% rownames_to_column("gene")
#     
#     ## Compute logFC
#     if(ncol(matSub) == 3) {
#       colnames(matSub)[str_detect(string = colnames(matSub), pattern = "DMSO")] <- "DMSO"
#       colnames(matSub)[str_detect(string = colnames(matSub), pattern = x)] <- "Treatment"
#       
#       matSub <- matSub %>% 
#         mutate(logFC = Treatment - DMSO) %>%
#         select(gene, logFC)
#     } else {
#       matSub <- data.frame(gene = matSub$gene, logFC = NA)
#     }
#     colnames(matSub)[str_detect(string = colnames(matSub), pattern = "logFC")] <- x
#     matSub %>% column_to_rownames("gene")
#   }) %>% bind_cols()
# })
# names(diffMat.list) <- compDrug
# 
# 
# diffMat.list
# diffMat.list[c(6)]
# 
# "lysosome_related", "RAB", "RAS", "autophagy_docking", "ULK1", "PI3K", "BLN", "auto_ubi"
# 
# path.info.sub <- path.info %>%
#   mutate(pathway = ifelse(pathway %in% c("NFKB", "NFKB_Chemo", "NFKB_GF", "NFKB_TF", "NFKB_Apo", "p65", "p52", "TNF"), "TNF-NFkB", pathway)) %>%
#   mutate(pathway = ifelse(pathway %in% c("auto_pos"), "Autophagy", pathway))
# 
# tf.sub <- data.frame(gene = target.df %>% filter(motif_name %in%c("HIF1A", "STAT3", "TFEB", "YY1", "PPARA", "PPARG", "SREBF1")) %>% pull(targets), pathway = "mTOR", source = "tftargets")
# 
# path.info.sub <- rbind(path.info.sub, tf.sub)
# 
# ## Significant genes
# sigGenes <- de.res %>% filter(p.adj < 0.1) %>%
#   #filter(Treatment == "Nutlin-3a") %>%
#   filter(logFC > 0.25 | logFC < -0.25) %>%
#   pull(gene) %>% unique()
# 
# 
# ## Matrix of logFC
# drugSub <- c("Everolimus")
# 
# drugSub <- de.res$Treatment %>% unique()
# 
# padjMat <- de.res %>% filter(Treatment %in% drugSub) %>%
#   filter(gene %in% sigGenes) %>%
#   #filter(logFC > 0.25 | logFC < -0.25) %>%
#   mutate(p.adj = -log10(p.adj)) %>%
#   mutate(p.adj = ifelse(logFC < 0, -1*p.adj, p.adj)) %>%
#   select(gene, Treatment, p.adj) %>%
#   pivot_wider(names_from = "Treatment", values_from = "p.adj") %>%
#   column_to_rownames("gene")
# 
# padjMat
# 
# ## unsure
# pHeat <- lapply(c("mTORC1", "TNF-NFkB", "TLR", "TP53up", "Autophagy"), function(x) {
# 
#   geneSelec <- path.info.sub %>% filter(pathway %in% c(x)) %>% select(gene) %>% unlist() %>% as.vector() %>% unique()  
# 
#   de.res <- de.res %>% filter(Treatment %in% drugSub)
#   
#   geneSig <- de.res %>% filter(Symbol %in% geneSelec) %>% pull(gene)
#   geneSig <- intersect(geneSig, rownames(padjMat))
#   
#   Heatmap(t(padjMat[geneSig, ]), show_row_names = TRUE,
#         cluster_rows = TRUE, border = TRUE, #column_names_gp = gpar(fontsize = 8),
#         show_column_names = FALSE, 
#         cluster_columns = TRUE, #row_names_gp = gpar(fontsize = 7),
#         clustering_method_rows = "ward.D2", #column_title = paste0(z),
#         clustering_method_columns = "ward.D2",
#         col=colorRamp2(c(-4, 0, 4), colors = c("#1976D2", "white", "#F44336")), name = "log10(p.adj)",
#         column_title = paste0(x))
# })
# pHeat[[1]] + pHeat[[2]] + pHeat[[3]] + pHeat[[4]] + pHeat[[5]] #+ pHeat[[6]] #+ pHeat[[7]] + pHeat[[8]] + pHeat[[9]] + pHeat[[10]] + pHeat[[11]]
# 
# 
# pHeat <- lapply(c("HIF1A", "STAT3", "TFEB", "YY1", "PPARA", "PPARG", "SREBF1"), function(x) {
# 
#   geneSelec <- target.df %>% filter(motif_name %in% x) %>% pull(targets) 
# 
#   de.res <- de.res %>% filter(Treatment %in% drugSub, p.adj < 0.1) %>%
#     filter(logFC > 0.25 | logFC < -0.25)
#   
#   geneSig <- de.res %>% filter(Symbol %in% geneSelec) %>% pull(gene)
#   geneSig <- intersect(geneSig, rownames(padjMat))
#   
#   Heatmap(t(padjMat[geneSig, ]), show_row_names = TRUE,
#         cluster_rows = TRUE, border = TRUE, #column_names_gp = gpar(fontsize = 8),
#         show_column_names = FALSE, 
#         cluster_columns = TRUE, #row_names_gp = gpar(fontsize = 7),
#         clustering_method_rows = "ward.D2", #column_title = paste0(z),
#         clustering_method_columns = "ward.D2",
#         col=colorRamp2(c(-4, 0, 4), colors = c("#1976D2", "white", "#F44336")), name = "log10(p.adj)",
#         column_title = paste0(x))
# })
# pHeat[[1]] + pHeat[[2]] + pHeat[[3]] + pHeat[[4]] + pHeat[[5]] + pHeat[[6]] + pHeat[[7]] #+ pHeat[[8]] + pHeat[[9]] + pHeat[[10]] + pHeat[[11]]
# 
# 
# de.r
# pHeat[[4]]
# 
# ## Make heatmap
# lapply(compDrug, function(x) {
#   message(x)
#   plotMat <- diffMat.list[[x]]
#   
#   ## Remove missing patients
#   plotMat <- plotMat[, complete.cases(t(plotMat))]
#   
#   pHeat <- Heatmap(plotMat, show_row_names = FALSE,
#           cluster_rows = TRUE, border = TRUE, #column_names_gp = gpar(fontsize = 8),
#           cluster_columns = TRUE, #row_names_gp = gpar(fontsize = 7),
#           clustering_method_rows = "ward.D2", #column_title = paste0(z),
#           clustering_method_columns = "ward.D2",
#           col=colorRamp2(c(-4, 0, 4), colors = c("#1976D2", "white", "#F44336")), name = "log2FC",
#           column_title = paste0("logFC Untreated vs ", x))
#   
#   png(file=paste0(opt$plot, x, "_logFC_heatmap.png"),
#       height = 10, width = 15, res = 600, units = "cm", bg = "transparent")
#   draw(pHeat, background = "transparent")
#   dev.off()
#   pHeat
# })
# 
# 
# 
# lapply(c("mTORC1", "NFKB", "TP53up", "TLR")[[4]], function(x) {
# 
#   geneSelec <- path.info.sub %>% filter(pathway %in% c(x)) %>% select(gene) %>% unlist() %>% as.vector() %>% unique()  
# 
#   de.res %>% filter(Symbol %in% geneSelec) %>%
#     mutate(pSign = -log10(p.adj) * sign(logFC), 
#            logFCScale = abs(logFC)) %>%
#     ggplot(aes(x = Treatment, y = pSign, size = logFCScale)) +
#     geom_jitter(height = 0, shape = 21)
# })
# 
# lapply(c("mTORC1", "NFKB", "TP53up", "TLR")[[2]], function(x) {
# 
#   geneSelec <- path.info.sub %>% filter(pathway %in% c(x)) %>% select(gene) %>% unlist() %>% as.vector() %>% unique()  
# 
#   de.res %>% #filter(Symbol %in% geneSelec) %>%
#     mutate(diffCol = ifelse(Symbol %in% geneSelec, 1, 0)) %>%
#     ggplot(aes(x = Treatment, y = logFC, size = -log10(p.adj), fill = diffCol)) +
#     geom_jitter(height = 0, shape = 21, width = 0.25) +
#     geom_hline(yintercept = 0)
# })

```

#### GSEA
```{r fig.height= 7.5, fig.width=25}

go.df.merge <- mclapply(c("Birinapant", "Motolimod", "Selinexor", "BafilomycinA1", "Dacinostat", "Pomalidomide", "Ibrutinib", 
                        "Everolimus", "Venetoclax", "Ruxolitinib", "Nutlin-3a"), mc.cores = ncores, 
                      function(compTreat) {
    message(paste("Running gprofiler for", compTreat, sep = " "))
    sigUp <- de.res %>% filter(logFC > 0.25, Treatment == compTreat, 
                               #!gene %in% c(drugOther), 
                               p.adj <= 0.1) %>% arrange(desc(logFC))
    #sigUp <- de.res %>% filter(celltype == compType, logFC > 0.25, Treatment == compTreat, p.adj <= 0.05) %>% arrange(PValue)
    
    sigUp
    if(nrow(sigUp) > 1) {
      gostresUp <- gost(query = sigUp$gene,
                        organism = "hsapiens", ordered_query = TRUE,
                        multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
                        measure_underrepresentation = FALSE, evcodes = FALSE,
                        user_threshold = 0.1, correction_method = "fdr",
                        domain_scope = "annotated", custom_bg = NULL,
                        numeric_ns = "", sources = NULL, as_short_link = FALSE)
      gostresUp$result

      go.df.up <- gostresUp$result %>% arrange(p_value)
      go.df.up$rank <- 1:nrow(go.df.up)
      go.df.up$type <- "Up"
    } else {
      go.df.up <- data.frame(p_value = NA, rank = NA, type = "Up", source = NA, term_name = NA, 
                               intersection_size = 10)
    }

    sigDown <- de.res %>% filter(logFC < -(0.25), Treatment == compTreat,
                                 #!gene %in% c(drugOther),
                                  p.adj <= 0.1) %>% arrange(logFC)
    #sigDown <- de.res %>% filter(celltype == compType, logFC < -(0.25), Treatment == compTreat, p.adj <= 0.05) %>% arrange(PValue)
    if(nrow(sigDown) > 1) {
    gostresDown <- gost(query = sigDown$gene,
                        organism = "hsapiens", ordered_query = TRUE,
                        multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
                        measure_underrepresentation = FALSE, evcodes = FALSE,
                        user_threshold = 0.1, correction_method = "fdr",
                        domain_scope = "annotated", custom_bg = NULL,
                        numeric_ns = "", sources = NULL, as_short_link = FALSE)

    gostresDown$result

    go.df.down <- gostresDown$result %>% arrange(p_value)
    go.df.down$rank <- 1:nrow(go.df.down) * -1
    go.df.down$type <- "Down"
    } else {
      go.df.down <- data.frame(p_value = NA, rank = NA, type = "Down", source = NA, term_name = NA, 
                               intersection_size = 10)
    }

    go.df.merge <- rbind(go.df.up, go.df.down)
    go.df.merge$Treatment <- compTreat

    go.df.merge <- go.df.merge %>% filter(!intersection_size < 10) %>% 
      filter(!source %in% c("CORUM", "TF", "MIRNA")) %>%
      mutate(p.adj = p.adjust(p_value, method = "BH"))
    
    if(nrow(go.df.merge) < 1) {
      go.df.merge <- data.frame(p_value = NA, rank = NA, type = "Down", source = NA, term_name = NA, 
                               intersection_size = 0)
    } else {
      go.df.merge
    }
}) %>% bind_rows() 

#### Show number of enriched gene sets
sigTab <- go.df.merge %>%
  dplyr::count(Treatment, type) %>%
  dplyr::rename(Direction = type) %>%
  mutate(n = ifelse(Treatment %in% c("Ruxolitinib", "Venetoclax"), 0, n)) %>%
  mutate(Treatment = factor(Treatment, levels = drugOrder), 
         n = ifelse(Direction == "Down", n * -1, n)) %>%
  filter(!is.na(Treatment))

naTab <- data.frame(Treatment = c("Venetoclax", "Venetoclax", "Selinexor"), 
                    Direction = c("Down", "Up", "Down"), 
                    n = c(0, 0, 0))
sigTab <- rbind(sigTab, naTab)

p <- sigTab %>%
  ggplot(aes(x = Treatment, y = n, fill = Direction)) +
  geom_bar(stat = "identity", colour = "black") +
  geom_text(aes(label = n), data = sigTab[sigTab$Direction == "Up", ], size = 5, nudge_y = 200) +
  geom_text(aes(label = n), data = sigTab[sigTab$Direction == "Down", ], size = 5, nudge_y = -200) +
  xlab("") + ylab("Number of Gene Sets") +
  ggtitle("Significantly Enriched Gene Sets") +
  lgd + scale_fill_manual(values = c("Up" = "#F44336", "Down" = "#1976D2")) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1), 
        plot.title = element_text(size = 21, vjust = 1), 
        axis.text = element_text(size = 17.5), 
        text = element_text(size = 22.5), 
        panel.border = element_rect(linewidth = 1))
p
ggsave(plot = p, filename = paste0(opt$plot, "n_gsea.png"), 
       height = 17.5, width = 22, units = "cm")

go.df.merge %>%
  filter(Treatment %in% c("Everolimus")) %>%
  ggplot(aes(x = Treatment, y = intersection_size)) +
  geom_jitter()

## Exclude all terms that are up and downregulated by a single-drug
dblTerm <- go.df.merge %>% 
  filter(!intersection_size < 10) %>%
  filter(!source %in% c("CORUM", "TF", "MIRNA")) %>%
  mutate(term_name = paste0(source, "_", term_name)) %>% unique() %>%
  #filter(term_name %in% unique(termTab$term_name)) %>%
  dplyr::count(term_name, Treatment) %>%
  filter(n > 1) %>%
  pull(term_name) %>% unique()

## Keep the top20 of each compound
nTop <- 5
termTab <- lapply(c("Selinexor", "Nutlin-3a", "Venetoclax", "Birinapant", "Ibrutinib", "Motolimod", 
                    "Dacinostat", "Pomalidomide", "Everolimus"), function(y) {
  termTab <- lapply(c("Up", "Down"), function(z) {
    message(y)
    message(z)
      
    gseaTab <- go.df.merge %>% filter(type == z, Treatment == y) %>% 
      filter(!intersection_size < 10) %>%
      filter(!source %in% c("CORUM", "TF", "MIRNA")) %>%
      mutate(term_name = paste0(source, "_", term_name)) %>% unique() %>%
      filter(!term_name %in% dblTerm)
      
    gseaTab <- gseaTab[1:nTop, ] # get only top enrichments
      
    #termOrder <- gseaTab %>% arrange(desc(p_value)) %>% pull(term_name)
      
    gseaTab <- gseaTab %>%
      mutate(p.adj = p.adjust(p_value, method = "BH")) 
  }) %>% bind_rows()
}) %>% bind_rows()

gseaTab <- go.df.merge %>% 
  filter(!intersection_size < 10) %>%
  filter(!source %in% c("CORUM", "TF", "MIRNA")) %>%
  mutate(term_name = paste0(source, "_", term_name)) %>% unique() %>%
  filter(term_name %in% unique(termTab$term_name)) %>%
  filter(!term_name %in% dblTerm)

plotMat <- gseaTab %>%
  #filter(Treatment %in% c("Motolimod", "Ibrutinib")) %>%
  #filter(source == "REAC") %>%
    #filter(term_name %in% termSub[[1]]) %>%
    mutate(p.adj = -log10(p.adj), p.adj = ifelse(type == "Down", p.adj * -1, p.adj)) %>%
    select(term_name, Treatment, p.adj) %>% unique() %>%
    pivot_wider(names_from = "Treatment", values_from = "p.adj") %>%
    column_to_rownames("term_name")

plotMat[is.na(plotMat)] <- 0
  
# Heatmap(plotMat, 
#         cluster_rows = TRUE, border = TRUE, #column_names_gp = gpar(fontsize = 8),
#         cluster_columns = TRUE, #row_names_gp = gpar(fontsize = 7),
#         clustering_method_rows = "ward.D2", #column_title = paste0(z),
#         clustering_method_columns = "ward.D2",
#         col=colorRamp2(c(-4, 0, 4), colors = c("#1976D2", "white", "#F44336")), name = "log2FC"#,
#         #column_title = paste0("logFC Untreated vs ", x)
#         )

geneOrder <- rownames(plotMat)[hclust(dist(plotMat), method = "ward.D2")$order]

p <- gseaTab %>%
  mutate(term_name = factor(term_name, levels = geneOrder), 
         intersection_size = ifelse(intersection_size > 200, 200, intersection_size)) %>%
  #mutate(term_name = str_wrap(string = term_name, width = 45)) %>%
  #filter(Treatment %in% c("Motolimod", "Ibrutinib")) %>%
  #filter(source == "REAC") %>%
    #filter(term_name %in% termSub[[1]]) %>%
  mutate(p.adj = -log10(p.adj), p.adj = ifelse(type == "Down", p.adj * -1, p.adj)) %>%
  ggplot(aes(x = term_name, y = Treatment)) +
  geom_point(aes(size = intersection_size, fill = p.adj), shape = 21) +
  xlab("") + ylab("") + ggtitle("Top Enriched Gene Sets") +
  lgd +
  theme(axis.text = element_text(size = 10), 
        axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 8.5), 
        panel.background = element_rect(linewidth = 1, fill = "transparent")) +
  scale_size_continuous(name = "Number of \nOverlapping Genes", 
                        breaks = seq(0, max(gseaTab$intersection_size), by = 50), 
                        limits = c(1, 200), range = c(0.5,7.5)) +
  scale_fill_gradient2(low = "#1976D2", high = "#F44336", limits=c(-4, 4), oob=squish, 
                       name = "-Log10(p.adj) \nwith Direction")
  
p

ggsave(plot = p, filename = paste0(opt$plot, "top_genesets_dotplot.png"), 
       height = 19.5, width = 50, units = "cm")

# 
# 
# lapply(c("Pomalidomide", "Nutlin-3a", "Birinapant", "Motolimod", "Selinexor", "Dacinostat"), function(x) {
#     message(x)
#   hyper.df <- hyper.df <- go.df.merge %>% 
#       filter(!source %in% c("CORUM", "TF", "MIRNA"), !intersection_size < 10, Treatment %in% x) %>%  
#         mutate(term_name = paste0(source, "_", term_name)) %>% unique() %>%
#         filter(!term_name %in% dblTerm)
#     
#   lapply(unique(hyper.df$type), function(y) {
#     hyper.df <- go.df.merge %>% 
#       filter(!source %in% c("CORUM", "TF", "MIRNA"), !intersection_size < 10, type == y, Treatment %in% x) %>%
#       arrange(p.adj) %>%
#       filter(p.adj < 0.1) %>%
#       mutate(p.adj = -log10(p.adj))
#     
#     hyper.df$rank <- 1:nrow(hyper.df)
#       
#     ## Get the top 5 enriched TPs
#     toptf <- hyper.df %>% filter(rank <= 10) %>%
#       pull(term_name)
#     
#     if(y == "Up") {col.point <- "#F44336"}
#     if(y == "Down") {col.point <- "#1976D2"}
#     
#     hyper.df %>%
#       ggplot(aes(x = rank, y = p.adj)) +
#       geom_point() +
#       geom_point(data = hyper.df[hyper.df$rank %in% 1:10, ], fill = col.point, size = 2.5, shape = 21) +
#       xlab("Rank") + ylab("-Log10(p.adj)") + ggtitle(paste0(x, " - Gene Set Enrichment ", y)) +
#       geom_label_repel(aes(label = term_name), data = hyper.df[hyper.df$rank %in% 1:10, ], 
#                        size = 5, max.overlaps = getOption("ggrepel.max.overlaps", default = 30),
#                        segment.linetype = 2, force = 20, alpha = 0.8) +
#       lgd +
#       theme_classic() +
#       theme(plot.title = element_text(hjust = 0.5),
#             legend.position = "none",
#             text = element_text(size = 17.5),
#             axis.text = element_text(size = 15),
#             panel.background = element_rect(fill = "transparent",colour = NA),
#             plot.background = element_rect(fill = "transparent",colour = NA),
#             legend.background = element_rect(fill='transparent'),
#             strip.background = element_blank())
# 
#   })
# })
# 


# nTop <- 5
# lapply(c("BafilomycinA1", "Birinapant", "Nutlin-3a", "Selinexor", "Ibrutinib", "Motolimod", "Everolimus", 
#          "Dacinostat", "Pomalidomide")[[4]], function(y) {
#   message(y)
# 
#   gseaTab <- go.df.merge %>% filter(Treatment == y) %>%
#     filter(!source %in% c("CORUM", "TF", "MIRNA")) %>%
#     mutate(term_name = paste0(source, "_", term_name)) %>% unique() %>%
#     mutate(p.adj = -log10(p.adj), p.adj = ifelse(type == "Down", p.adj * -1, p.adj))
# 
#   termSub <- gseaTab %>%
#     dplyr::count(term_name) %>%
#     filter(n == 1) %>%
#     pull(term_name)
#   
#   gseaTab <- gseaTab %>% filter(term_name %in% termSub) %>%
#     arrange(desc(p.adj)) %>%
#     mutate(term_name = str_wrap(string = term_name, width = 45))
# 
#   gseaTab.up <- gseaTab[gseaTab$p.adj > 0, ]
#   rownames(gseaTab.up) <- 1:nrow(gseaTab.up)
#   
#   if(nrow(gseaTab.up) < nTop) {
#     gseaTab.up
#   
#   } else {
#     gseaTab.up <- gseaTab.up[1:nTop, ]
#   }
#   
#   
#   gseaTab.down <- gseaTab[gseaTab$p.adj < 0, ]
#   rownames(gseaTab.down) <- 1:nrow(gseaTab.down)
#   if(nrow(gseaTab.down) < nTop) {
#     gseaTab.down
#   
#   } else {
#     gseaTab.down <- gseaTab[c(c(nrow(gseaTab.down) - nTop):nrow(gseaTab.down)), ]
#   }
#   
#   gseaTab <- rbind(gseaTab.up, gseaTab.down) %>%
#     arrange(p.adj)
#   
#   #nrow(gseaTab)
#   # gseaTab
#   # if(max(gseaTab$p.adj) > 25) {
#   #   x.max <- max(gseaTab$p.adj)
#   # } else if(y == "Birinapant") {
#   #   x.max <- 35
#   # } else {
#   #   x.max <- 25
#   # }
#   # 
#   # if(min(gseaTab$p.adj) < -5) {
#   #   x.min <- min(gseaTab$p.adj)
#   # } else {
#   #   x.min <- -5
#   # }
#   # 
#   # p <- gseaTab %>%
#   #   mutate(term_name = factor(term_name, levels = term_name)) %>%
#   #   ggplot(aes(x = p.adj, y = term_name)) +
#   #   geom_bar(stat = "identity", colour = "black", alpha = 0.5, aes(fill = type)) +
#   #   #geom_vline(xintercept = -log10(0.1), alpha = 0.8, linetype = "dashed") +
#   #   #geom_vline(xintercept = log10(0.1), alpha = 0.8, linetype = "dashed") +
#   #   geom_text(aes(label = term_name, y = term_name, x = 0), hjust = 0, size = 3.75, data = gseaTab[gseaTab$p.adj > 0, ]) +
#   #   geom_text(aes(label = term_name, y = term_name), hjust = 0, size = 3.75, data = gseaTab[gseaTab$p.adj < 0, ]) +
#   #   xlab("-Log10(p.adj)") + ylab("") + #ggtitle(paste0("Top enriched gene-set: ", z)) +
#   #   ggtitle(paste0(y)) +
#   #   xlim(x.min, x.max) +
#   #       lgd + theme(axis.text.y = element_blank(),
#   #                   axis.ticks.y = element_blank(),
#   #                   text = element_text(size = 22.5),
#   #                   panel.border = element_rect(colour = "black", fill=NA, size=1.5),
#   #                   strip.background = element_blank(),
#   #                   panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
#   #                   panel.background = element_rect(fill = "transparent",colour = NA),
#   #                   plot.background = element_rect(fill = "transparent",colour = NA),
#   #                   legend.background = element_rect(fill='transparent'),
#   #                   legend.position = "none") +
#   #     scale_fill_manual(values = c("Down" = "#1976D2", "Up" = "#F44336"))
#   #     ggsave(plot = p,
#   #            file = paste0(opt$plot, "GSEA/", y, "_top.png"),
#   #            height = 15, width = 12.5, units = "cm")
#   #    p
#   # 
# })


## Plot top enriched pathwaysgo.df.merge
# nTop <- 20
# 
# lapply(c("BafilomycinA1", "Ibrutinib", "Birinapant", "Selinexor", "Pomalidomide", "Dacinostat", 
#          "Everolimus", "Motolimod", "Nutlin-3a"), function(y) {
#            
#   gseaTab <- go.df.merge %>% filter(Treatment == y) %>% filter(!source %in% c("CORUM", "TF", "MIRNA")) %>%
#         mutate(term_name = paste0(source, "_", term_name)) %>% unique()
#   
#   message(y)
#      lapply(unique(gseaTab$type), function(z) {
#       message(z)
#       
#       gseaTab <- gseaTab %>% filter(type == z)
#       
#       gseaTab <- gseaTab[1:nTop, ] # get only top enrichments
#       
#       termOrder <- gseaTab %>% arrange(desc(p_value)) %>% pull(term_name)
#       
#       gseaTab <- gseaTab %>%
#         mutate(term_name = factor(term_name, levels = termOrder), 
#                p.adj = p.adjust(p_value, method = "BH"))
#       
#     if(-log10(min(gseaTab$p.adj, na.rm = TRUE)) > 10) {
#       x.max <- -log10(min(gseaTab$p.adj, na.rm = TRUE))
#     } else {
#       x.max <- 10
#     }
#     if(y %in% c("Birinapant", "Selinexor", "Pomalidomide")) {
#       x.max <- 3
#     }
#     if(y %in% c("Pomalidomide")) {
#       x.max <- 2
#     }
# 
# 
#       
#       p <- gseaTab %>%
#         ggplot(aes(x = -log10(p.adj), y = term_name)) +
#         geom_bar(stat = "identity", colour = "black", alpha = 0.5, aes(fill = type)) +
#         geom_vline(xintercept = -log10(0.1), alpha = 0.8, linetype = "dashed") +
#         geom_text(aes(label = term_name, y = term_name, x = 0.25), hjust = 0, size = 4.25) +
#         xlab("-Log10(p.adj)") + ylab("") + #ggtitle(paste0("Top enriched gene-set: ", z)) +
#         ggtitle(paste0(y, " - ", z)) +
#         xlim(0, x.max) +
#         lgd + theme(axis.text.y = element_blank(), 
#                     axis.ticks.y = element_blank(), 
#                     text = element_text(size = 22.5),
#                     panel.border = element_rect(colour = "black", fill=NA, size=1.5), 
#                     strip.background = element_blank(),
#                     panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
#                     panel.background = element_rect(fill = "transparent",colour = NA),
#                     plot.background = element_rect(fill = "transparent",colour = NA),
#                     legend.background = element_rect(fill='transparent'), 
#                     legend.position = "none") +
#         scale_fill_manual(values = c("Down" = "#1976D2", "Up" = "#F44336"))
#         #scale_fill_npg() +
#   #guides(fill=guide_legend(title="Source"))
#       ggsave(plot = p,
#              file = paste0(opt$plot, "GSEA/", y, "_", z, "_top_", nTop, ".png"),
#              height = 22.5, width = 19, units = "cm")
#       p
#     })
# })


```

#### Hypergeometric test for transcription factors
```{r}
# treatSub <- c("Nutlin-3a", "Selinexor", "Birinapant", "Ibrutinib", "Pomalidomide", 
#               "Motolimod", "Everolimus", "BafilomycinA1", "Venetoclax", "Dacinostat", 
#               "Ruxolitinib")
# 
# ## Run hypergeometric test
# hyper.df <- lapply(treatSub, function(x) {
#   ## Limit the number of targets
#   motif.df <- motif.df %>% filter(n_targets < maxtargets)
#   message("Keeping ", nrow(motif.df), " TFs with less than ", maxtargets)
# 
#   ## Get the transcription factor targets
#   lapply(c("Up", "Down"), function(y) {
#     df.hub <- de.res %>% filter(p.adj < 0.05, Treatment == x)
#     
#     if(y == "Up") {df.hub <- df.hub %>% filter(logFC > 0.25)}
#     if(y == "Down") {df.hub <- df.hub %>% filter(logFC < -0.25)}
#     
#     mclapply(unique(motif.df$motif_name), mc.cores = ncores, 
#              function(z) {
#       message(z)
#       tf.targets <- target.df %>% filter(motif_name == z) %>% pull(targets)
# 
#       size.module <- nrow(df.hub)
#       size.genome <- length(unique(de.res$Symbol))
#       size.genome <- nrow(sce)
#       n.targets <- length(tf.targets)
#       n.over <- intersect(df.hub$Symbol, tf.targets) %>% length()
# 
#       p.val <- phyper(q = n.over - 1, # number of white balls drawn
#                       m = n.targets, # number white balls in urn
#                       n = size.genome - n.targets, # number black balls in urn
#                       k = size.module, # number of draws
#                       lower.tail= FALSE)
# 
#       df <- data.frame(Treatment = x, dir = y, tf = z, p.val = p.val, n.over = n.over,
#                        n.targets = n.targets, size.module = size.module,
#                        size.genome = size.genome)
#       #df %>% filter(n.over != 0)
#     }) %>% bind_rows()
#   }) %>% bind_rows() %>% mutate(p.adj = p.adjust(p.val, method = "BH"))
# })# %>% bind_rows()
# 
# names(hyper.df) <- treatSub
# 
# lapply(c("Up", "Down"), function(x) {
#   plotMat <- hyper.df %>% bind_rows() %>%
#     filter(dir == x) %>%
#     select(Treatment, tf, p.adj) %>%
#     mutate(p.adj = as.numeric(p.adj), 
#            p.adj = -log10(p.adj)) %>%
#     pivot_wider(names_from = "tf", values_from = "p.adj") %>%
#     column_to_rownames("Treatment")
# 
#   ## Filter for TFs that are significanlty targeted by at least one drug
#   tf.sub <- hyper.df %>% bind_rows() %>% filter(dir == x) %>% filter(p.adj < 0.001) %>% pull(tf) %>% unique()
#   message("Plotting ", length(tf.sub), " TFs")
#   
#   plotMat <- plotMat[, tf.sub]
#   plotMat <- scale(plotMat)
#   
#   ## Plot
#   set.seed(100)
#   pHeat <- Heatmap(plotMat[, tf.sub], 
#           clustering_method_rows = "ward.D2", 
#           clustering_method_columns = "ward.D2", border = TRUE, 
#           column_title = paste0("Enriched Transcription Factors Among DEGs (FDR < 0.1%) ", x), 
#           show_column_names = FALSE, heatmap_legend_param = list(title_gp = gpar(fontsize = 12.5)),
#           col=colorRamp2(c(-4, 0, 4), colors = c("#1976D2", "white", "#F44336")),
#           #col=colorRamp2(c(0, 50), colors = c("grey90", "#F44336")), 
#           name = "-Log10(p.adj)")
#   
#   #lgd <- Legend(title = "ward.D2", legend_gp = gpar(fontsize = 20))
#   # 
#   # png(file=paste0(opt$plot, x, "_TF_heatmap.png"),
#   #     height = 10, width = 20, res = 600, units = "cm", bg = "transparent")
#   # draw(pHeat, background = "transparent"
#   #      #, heatmap_legend_list = lgd
#   #      )
#   # dev.off()
#   pHeat
# })
# hyper.df %>% bind_rows() %>% 
#   filter(Treatment == "Pomalidomide", ) %>%
#   arrange(p.adj) %>%
#   filter(p.adj < 0.1) %>%
#   mutate(p.adj = -log10(p.adj))
# 
# lapply(c("Pomalidomide", "Nutlin-3a", "Birinapant", "Motolimod", "Selinexor", "Dacinostat", "Ibrutinib", "Everolimus"), function(x) {
#   lapply(c("Up", "Down"), function(y) {
#     message(x)
#     hyper.df <- hyper.df %>% bind_rows() %>% 
#       filter(Treatment == x, dir == y) %>%
#       arrange(p.adj) %>%
#       filter(p.adj < 0.1) %>%
#       mutate(p.adj = -log10(p.adj))
#     
#     hyper.df$rank <- 1:nrow(hyper.df)
#       
#     ## Get the top 5 enriched TPs
#     toptf <- hyper.df %>% filter(rank <= 10) %>%
#       pull(tf)
#     
#     if(x == "Pomalidomide") {
#       toptf <- c(toptf, "IKZF1", "IKFZ3", "BRD7", "ARID2", "IRF4", "MYC")
#     }
# 
#     if(x %in% c("Nutlin-3a", "Selinexor")) {
#       toptf <- c(toptf, "TP53")
#     }
#     if(x %in% c("Birinapant", "Motolimod", "BafilomycinA1", "Ibrutinib")) {
#       toptf <- c(toptf, "NFKB", "NFKB1", "NFKB2", "REL", "RELA", "RELB")
#     }
#         
#     if(y == "Up") {col.point <- "#F44336"}
#     if(y == "Down") {col.point <- "#1976D2"}
#     
#     p <- hyper.df %>%
#       ggplot(aes(x = rank, y = p.adj)) +
#       geom_point() +
#       geom_point(data = hyper.df[hyper.df$tf %in% toptf, ], fill = col.point, size = 2.5, shape = 21) +
#       xlab("Rank") + ylab("-Log10(p.adj)") + ggtitle(paste0(x, " - TF Enrichment (", y, ")")) +
#       geom_label_repel(aes(label = tf), data = hyper.df[hyper.df$tf %in% toptf, ], 
#                        max.overlaps = getOption("ggrepel.max.overlaps", default = 30), 
#                        segment.linetype = 2, force = 20, alpha = 0.8) +
#       lgd +
#       theme_classic() +
#       theme(plot.title = element_text(hjust = 0.5),
#             legend.position = "none",
#             text = element_text(size = 17.5),
#             axis.text = element_text(size = 15),
#             panel.background = element_rect(fill = "transparent",colour = NA),
#             plot.background = element_rect(fill = "transparent",colour = NA),
#             legend.background = element_rect(fill='transparent'),
#             strip.background = element_blank())
# 
#     ggsave(plot = p, filename = paste0(opt$plot, x, "_", y, "_tf_enrichment.png"),
#            height = 12.5, width = 15, units = "cm")
#     
#     p
#   })
# })
# 
# lapply(c("Pomalidomide"), function(x) {
#   hyper.df <- hyper.df %>% bind_rows() %>% 
#     filter(Treatment == x) %>%
#     arrange(p.adj) %>%
#     filter(p.adj < 0.1) %>%
#     mutate(p.adj = -log10(p.adj), 
#            p.adj = ifelse(dir == "Down", p.adj * -1, p.adj)) %>%
#     arrange(p.adj)
#     
#   hyper.df$rank <- 1:nrow(hyper.df)
#   
#     ## Get the top 5 enriched TPs
#     toptf <- hyper.df %>% filter(rank <= 10) %>%
#       pull(tf)
# 
#     if(x == "Pomalidomide") {
#       toptf <- c(toptf, "IKZF1", "IKZF2", "IKFZ3")
#     }
#     # 
#     # if(y == "Up") {col.point <- "#F44336"}
#     # if(y == "Down") {col.point <- "#1976D2"}
#     
#   hyper.df %>%
#     ggplot(aes(x = rank, y = p.adj)) +
#     geom_point() +
#     #geom_point(data = hyper.df[hyper.df$tf %in% toptf, ], fill = col.point, size = 2.5, shape = 21) +
#     xlab("") + ylab("-Log10(p.adj)") + ggtitle(paste0(x, " - TF Enrichment")) +
#     geom_label_repel(aes(label = tf), data = hyper.df[hyper.df$tf %in% toptf, ]) +
#     lgd +
#     theme(axis.ticks.x = element_blank(), 
#           axis.text.x = element_blank())
# })
# 
# 
# hyper.df %>% bind_rows() %>% filter(Treatment == "Nutlin-3a") %>%
#   arrange(p.adj)

```


#### Hallmarks
```{r}
# 
# 
# gseaTab <- lapply(c("Birinapant", "Motolimod", "Selinexor", "BafilomycinA1", "Dacinostat", 
#                       "Pomalidomide", "Ibrutinib", "Everolimus", "Nutlin-3a"), function(y) {
#     testTab <- de.res %>% filter(Treatment == y) %>% # subset toeach celltype & significant DEG
#       filter(p.adj < 0.1) %>%
#       filter(logFC < 0.25 | logFC > 0.25) %>%
#       select(logFC, Symbol) %>% column_to_rownames("Symbol") %>% 
#       unique() # make a df with logFC as rank and genes as rownames.
#     DEG.df <- runGSEA(testTab, gmtFile = "data/h.all.v2022.1.Hs.symbols.gmt", 
#                       GSAmethod = "page", nPerm = 1000)
#     DEG.df$Treatment <- y
#     DEG.df
# }) %>% bind_rows()
# 
# library(jyluMisc)
# colnames(gseaTab) <- c("Name", "genes_tot", "stat_dir", "p_dir_up", "p.adj.up", "p_dir_down", 
#                        "p.adj.down", "genes_up", "genes_down", "Treatment")
# 

```


#### Heatmap of DEGs
```{r}
# drugSub <- c("Birinapant", "Nutlin-3a", "Selinexor", "Ruxolitinib", "Venetoclax", "Ibrutinib", "Motolimod", "Everolimus", "Pomalidomide")
# 
# geneSub <- de.res %>% filter(p.adj < 0.1) %>%
#   filter(Treatment %in% drugSub ) %>%
#   filter(logFC > 0.25 | logFC < -0.25) %>%
#   pull(gene) %>% unique()
# 
# plotMat <- de.res %>%
#   filter(gene %in% geneSub) %>%
#   filter(Treatment %in% drugSub ) %>%
#   mutate(p.adj = -log10(p.adj) * sign(logFC)) %>%
#   select(Symbol, Treatment, p.adj) %>%
#   pivot_wider(names_from = "Treatment", values_from = "p.adj") %>%
#   column_to_rownames("Symbol")
# 
# 
# Heatmap(plotMat, show_row_names = FALSE, 
#         clustering_method_rows = "ward.D2", #column_title = paste0(z),
#         clustering_method_columns = "ward.D2",
#         col=colorRamp2(c(-4, 0, 4), colors = c("#1976D2", "white", "#F44336")), name = "log2FC")

```

#### Output session info
```{r pressure, echo=FALSE}
sessionInfo()
```
