---
title: "Figure 4"
author: "M. Pohly, K. Bach"
date: "`r Sys.Date()`"
output: html_document
---

## Load libraries
```{r warning=FALSE, message=FALSE}
library(tidyverse)
library(scran)
library(scater)
library(gprofiler2)
library(gridExtra)
library(ComplexHeatmap)
library(circlize)
library(edgeR)
library(readxl)
library(ggrepel)
library(tftargets)
library(network)
library(ggnet)
library(parallel)
library(BiocParallel)
library(ggsci)
library(airr)
library(numbat)
library(scales)

ncores <- 10
mcparam <- MulticoreParam(workers=ncores)
register(mcparam)

```

```{r}
opt <- list()
opt$pathinfo <- "misc/pathways_2.xlsx"
opt$scesc <- "data/submission/scRNA_complete.RDS"
opt$scescpat <- "data/submission/scRNA_per_pat_complete.RDS"
#opt$types <- "data/loop2020_celltypes.csv"
#opt$immtab <- "data/loop2020_table_heavy_light_contr_mut_burden_tcr.tsv"
opt$plot <- "plots/Fig4/"

## Set theme
lgd <-  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5), 
        text = element_text(size = 15),
        panel.background = element_rect(fill = "transparent",colour = NA),
        plot.background = element_rect(fill = "transparent",colour = NA), 
        legend.background = element_rect(fill='transparent',colour = NA), 
        strip.background = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())


```

## Load data
```{r}
## scRNA-Seq data
sce.sc <- readRDS(opt$scesc)

## single-cell cell types
types <- sce.sc[["types"]]

## Load tcr data
immTab <- sce.sc[["immTab"]]

sce.sc <- sce.sc[["SCE"]]

## sce per pat
sce.pat <- readRDS(opt$scescpat)

## Load pathway annotation
path.info <- read_excel(opt$pathinfo)
```


#### Overview scRNA-seq
```{r}
## Randomly sample 
set.seed(100)
sce.sc <- sce.sc[, sample(1:ncol(sce.sc), ncol(sce.sc))]

## Re-run UMAP
set.seed(100)
sce.sc <- runUMAP(sce.sc, n_neighbors = 15, min_dist = 0.75, name = "UMAP", BPPARAM = mcparam)

## Make plots
plotTab <- ggcells(sce.sc)[[1]]

## Cell type
p <- plotTab %>%
  left_join(types, by = "uniqueBarcode") %>%
  mutate(celltype = ifelse(celltype %in% c("malignant B-cell", "malignant T-cell"), Diagnosis_simple, celltype)) %>%
  ggplot(aes(x = UMAP.1, y = UMAP.2, fill = celltype)) +
  geom_point(shape = 21, size = 2.5, stroke = 0.25) +
  lgd + theme_void() +
  scale_fill_manual(values = c("MCL" = "#F44336", "T-PLL" = "#64B5F6",
                                "healthy T-cell" = "#0D47A1", "NK-cell" = "#AB47BC",
                                "healthy B-cell" = "#FFA000", "HCL" = "#FFD45F",
                               "T-LGL" = "#80DEEA",
                                "monocyte" = "#66BB6A")) +
  theme(legend.position = "none")
p

ggsave(plot = p, filename = paste0(opt$plot, "scrna_umap_celltypes.png"), 
       height = 15, width = 15, units = "cm")

## Treatment
p <- plotTab %>%
  left_join(types, by = "uniqueBarcode") %>%
  mutate(celltype = ifelse(celltype %in% c("malignant B-cell", "malignant T-cell"), Diagnosis_simple, celltype)) %>%
  ggplot(aes(x = UMAP.1, y = UMAP.2, fill = Treatment)) +
  geom_point(shape = 21, size = 2.5, stroke = 0.25) +
  lgd + theme_void() +
  scale_fill_manual(values = c("DMSO" = "grey70", "Birinapant" = "#AD1457", "Nutlin-3a" = "#FFA000", 
                                "Ibrutinib" = "#283593", "Everolimus" = "#43A047", 
                               "Selumetinib" = "#F8BBD0")) +
  theme(legend.position = "none")
p
ggsave(plot = p, filename = paste0(opt$plot, "scrna_umap_treatment.png"), 
       height = 15, width = 15, units = "cm")

## PatientID
p <- plotTab %>%
  left_join(types, by = "uniqueBarcode") %>%
  mutate(celltype = ifelse(celltype %in% c("malignant B-cell", "malignant T-cell"), Diagnosis_simple, celltype)) %>%
  ggplot(aes(x = UMAP.1, y = UMAP.2, fill = patientID)) +
  geom_point(shape = 21, size = 2.5, stroke = 0.25) +
  lgd + theme_void() +
  scale_fill_npg() +
  theme(legend.position = "none")
p
ggsave(plot = p, filename = paste0(opt$plot, "scrna_umap_patientID.png"), 
       height = 15, width = 15, units = "cm")


```


#### Show donut plots
```{r}

plotTab.vdj <- plotTab %>%
  left_join(types, by = "uniqueBarcode") %>%
  filter(celltype %in% c("malignant T-cell", "healthy T-cell")) %>%
  mutate(celltype = ifelse(celltype == "malignant T-cell", patientID, celltype), 
         celltype = ifelse(celltype == "healthy T-cell", "Healthy T-Cells", celltype))


plotTab.vdj[plotTab.vdj$celltype == "H371" , "celltype"] <- "T-PLL1"
plotTab.vdj[plotTab.vdj$celltype == "H431", "celltype"] <- "T-PLL2"
plotTab.vdj[plotTab.vdj$celltype == "H279", "celltype"] <- "T-PLL3"
plotTab.vdj[plotTab.vdj$celltype == "H501", "celltype"] <- "T-LGL1"

## Merge VDJ data with plotting info
plotTab.vdj <- plotTab.vdj %>% left_join(select(immTab, cell_id, clone_id, c_call, #mu_freq, sequence_alignment
                                                ), 
                                     by = c("uniqueBarcode" = "cell_id"))

## Calculate the number of clones per cell type
countTab <- plotTab.vdj %>%
  group_by(celltype) %>%
  select(celltype, clone_id) %>% unique() %>%
  dplyr::count(celltype) %>% mutate(n_clones = n) %>%
  select(-n)
countTab

## Calculate the size of each clone
countSize <- plotTab.vdj %>%
  filter(!is.na(clone_id)) %>%
  group_by(celltype) %>%
  select(celltype, clone_id) %>%
  dplyr::count(celltype, clone_id) %>% mutate(clone_size = n) %>%
  select(-n)
countSize

###################
# Donut charts of clonal composition
###################
## Calculate the size of each clone - don't subset to only cells with clone_id
# otherwise NA T-cells will not be in the plot later.
cloneCount <- plotTab.vdj %>% 
  group_by(celltype) %>%
  dplyr::count(clone_id) %>% ungroup()
cloneCount

## This script was written for a B-cell analysis. Therefore, some labels 
# might refer to B-cells, but we're looking at T-cells at the moment. 

cloneNum <- lapply(unique(plotTab.vdj$celltype), function(x) {
  subTab <- filter(plotTab.vdj, celltype == x) # subset to each patient
  countTab <- dplyr::count(subTab, clone_id)
  nclones <- countTab %>% nrow() # get the number of clones
  ncells_bcr <- nrow(subTab) # calculate the number of B or T cells per patient - important to subset plotTab to only B or T-cell clusters before doing this
  res <- data.frame(patID = x, nclones = nclones, ncells_bcr = ncells_bcr)
  res$type <- opt$subType
  res
}) %>% bind_rows()
cloneNum

df <- left_join(cloneCount, cloneNum, by = c("celltype" = "patID")) %>%
  group_by(celltype) %>% mutate(clone_ratio = n / ncells_bcr) %>% # divide the size of the clone 
  select(celltype, clone_id, clone_ratio, nclones) %>% unique() %>%
  mutate(ymax = cumsum(clone_ratio), ymin = c(0, head(ymax, n= -1)))
df

#### Overview over all patients/celltypes

pClone.Pat <- df %>% 
  filter(celltype %in% c("T-PLL1", "T-PLL2", "T-PLL3", "Healthy T-Cells")) %>%
  ggplot(aes(ymax = ymax, ymin = ymin, xmax=4, xmin=3, fill = clone_id)) +
  geom_rect(colour = "black", size = 0.25) +
  coord_polar(theta="y") +
  xlim(c(2, 4)) +
  # ggtitle(paste0("Clonal Composition per Patient - ", opt$subType)) +
  facet_wrap(. ~ celltype, nrow =  1) +
  theme_void() +
  theme(plot.title = element_text(hjust = 0.5),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 12.5),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_rect(fill = "transparent",colour = NA),
        plot.background = element_rect(fill = "transparent",colour = NA),
        legend.position = "none")

pClone.Pat

#### Output individual patients
subPat <- c("T-PLL1", "T-PLL2", "T-PLL3", "Healthy T-Cells")

lapply(subPat, function(subPat) {
  pClone <- df %>% filter(celltype == subPat) %>%
  ggplot(aes(ymax = ymax, ymin = ymin, xmax=4, xmin=3, fill = clone_id)) +
  geom_rect(colour = "black", size = 0.25) +
  coord_polar(theta="y") +
  xlim(c(2, 4)) +
  ggtitle(paste0(subPat)) +
  theme_void() +
  theme(plot.title = element_text(hjust = 0.5, size = 75),
        legend.position = "none")
  pClone
  
  ggsave(plot = pClone, filename = paste0(opt$plot, "tcr_comp_", subPat, ".png"), 
         height = 20, width = 20, units = "cm")
})


```

#### Perform differential gene expression testing - scRNA-Seq
```{r message=FALSE, warning=FALSE}
opt$subsample <- TRUE

compDrug <- c("Nutlin-3a", "Birinapant", "Ibrutinib", "Everolimus", "Selumetinib")

sce_sub <- sce.sc
plotTab <- ggcells(sce_sub)[[1]]

colData(sce_sub) <- colData(sce_sub) %>%
  data.frame() %>%
  left_join(types, by = "uniqueBarcode") %>%
  mutate(celltype = ifelse(celltype %in% c("malignant B-cell", "malignant T-cell"), Diagnosis_simple, celltype)) %>%
  mutate(celltype = ifelse(celltype %in% c("healthy T-cell"), "Healthy T-Cells", celltype)) %>%
  mutate(celltype = ifelse(celltype %in% c("NK-cell"), "NK-Cells", celltype)) %>%
  mutate(celltype = ifelse(celltype %in% c("monocyte"), "Monocytes", celltype)) %>%
  mutate(celltype = ifelse(patientID%in% c("H501") & celltype %in% c("MCL", "HCL", "T-PLL"), "dead cell", celltype)) %>%
  mutate(celltype = ifelse(patientID%in% c("H431", "H371", "H279") & celltype %in% c("MCL", "HCL", "T-LGL"), "dead cell", celltype)) %>%
  mutate(celltype = ifelse(patientID%in% c("H526", "H525") & celltype %in% c("MCL", "T-PLL", "T-LGL"), "dead cell", celltype)) %>%
  mutate(celltype = ifelse(patientID%in% c("H496", "H432", "P1029") & celltype %in% c("HCL", "T-PLL", "T-LGL"), "dead cell", celltype)) %>%
  DataFrame()


if(opt$subsample == TRUE) {
 cellSub <- c("HCL", "MCL", "T-PLL", "Healthy T-Cells") 
} else {
  cellSub <- c("HCL", "MCL", "T-PLL")
}

testTreat <- c("Birinapant", "Nutlin-3a", "Ibrutinib", "Everolimus", "Selumetinib")

countTab <- plotTab %>%
  left_join(types, by = "uniqueBarcode") %>%
  mutate(celltype = ifelse(celltype %in% c("malignant B-cell", "malignant T-cell"), Diagnosis_simple, celltype)) %>%
  mutate(celltype = ifelse(celltype %in% c("healthy T-cell"), "Healthy T-Cells", celltype)) %>%
  mutate(celltype = ifelse(celltype %in% c("NK-cell"), "NK-Cells", celltype)) %>%
  mutate(celltype = ifelse(celltype %in% c("monocyte"), "Monocytes", celltype)) %>%
  group_by(celltype, Treatment) %>%
  dplyr::count() %>%
  filter(celltype %in% cellSub, Treatment %in% testTreat) %>%
  arrange(n)

minCells <- countTab[1, ]$n
minCells

de.res.2020 <- lapply(testTreat, function(z) {
  message(paste0("Fitting model for ", z))
  lapply(cellSub, #mc.cores = ncores, 
         function(x) {
    message(paste0("Celltype ", x))
    sce.x <- sce_sub[, sce_sub$celltype == x]
    sce.x <- sce.x[, sce.x$Treatment %in% c("DMSO", testTreat)]
    message(ncol(sce.x), " cells")
    
    if(opt$subsample == TRUE) {
      message("Subsetting to ", minCells)
      
      sce.x <- do.call(cbind, lapply(unique(sce.x$Treatment), function(y) {
        sce.y <- sce.x[, sce.x$Treatment == y]
        
        if(ncol(sce.y) == minCells) {sce.y} else {
          set.seed(101)
          sce.y <- sce.y[,sample(1:ncol(sce.y), minCells)]
          sce.y
        }
        sce.y
      }))
    }
 
    summed <- summarizeAssayByGroup(sce.x, ids = colData(sce.x)[,c("Diagnosis_simple", "patientID", "Treatment")], 
                                    statistics = "sum", assay.type = "counts") # get raw counts and sum them up
    #  summed <- applySCE(sce.x, aggregateAcrossCells, ids=colData(sce.x)[,c("Diagnosis_simple", "patientID", "Treatment")])
    colData(summed)$Treatment <- factor(colData(summed)$Treatment, levels = c("DMSO", compDrug))
    y <- DGEList(assay(summed), samples = colData(summed))
    opt$subsample
    if(opt$subsample == TRUE) {
      discarded <- summed$ncells < 5
    } else {
      discarded <- summed$ncells < 30
    }
    y <- y[,!discarded]
    message(paste0("Discarding ", sum(discarded), " samples"))
    design <- model.matrix(~0 + factor(patientID) + Treatment, y$samples) 

    keep <- filterByExpr(y, design = design)
    y <- y[keep,]
    summary(keep)
    message(paste0("Keeping ", sum(keep), " genes"))
    y <- calcNormFactors(y)
    # head(design)
    y <- estimateDisp(y, design) # gene-wise dispersion (variance on top of poisson distribution)
    #plotBCV(y)
    fit <- glmQLFit(y, design, robust=TRUE) # QL way of fitting the model
    #plotQLDisp(fit)
    res <- glmQLFTest(fit, coef = paste0("Treatment", z))
    # #res <- glmQLFTest(fit, coef = "TreatmentNutlin-3a")
    summary(decideTests(res))
    resTab <- res$table
    resTab$gene <- rownames(resTab)
    resTab$celltype <- x
    resTab <- resTab #%>%
      #mutate(p.adj = ifelse(p.adj < 1e-12, 1e-12, p.adj))
    resTab$Treatment <- z
    resTab %>% arrange(desc(logFC)) %>% mutate(p.adj = p.adjust(PValue, method = "BH"))
  }) %>% bind_rows()
}) %>% bind_rows()

de.res.2020 <- de.res.2020 %>%
  left_join(select(data.frame(rowData(sce.sc)), ID, Symbol), c("gene" = "ID"))

```

#### Show number of DEGs
```{r fig.height= 7.5, fig.width= 15}
################################################################################
# Show number of DEGs
################################################################################

plotMat <- de.res.2020 %>%
  #filter(PValue < 0.05) %>%
  filter(p.adj < 0.1) %>%
  filter(logFC > 0.25 | logFC < -0.25) %>%
  mutate(dir = ifelse(logFC < 0, "Down", "Up")) %>%
  group_by(celltype, dir, Treatment) %>%
  dplyr::rename(Direction = dir) %>%
  dplyr::count()

plotMat <- plotMat %>%
  mutate(n = ifelse(Direction == "Down", n * -1, n)) 

p <- plotMat %>%
  #mutate(celltype = factor(celltype, levels = c("HCL", "MCL", "T-PLL & T-LGL", "T- & NK-cells"))) %>%
  mutate(celltype = factor(celltype, levels = c("HCL", "MCL", "T-PLL", "Healthy T-Cells"))) %>%
  ggplot(aes(x = celltype, y = n, fill = Direction)) +
  geom_bar(colour = "black", stat = "identity") +
  geom_hline(yintercept = 1) +
  geom_text(aes(label = n), nudge_y = 50, data = plotMat[plotMat$Direction == "Up", ], size = 5) +
  geom_text(aes(label = n), nudge_y = -50, data = plotMat[plotMat$Direction == "Down", ], size = 5) +
  xlab("") + ylab("Number of Genes") + ggtitle("DEGs per Cell Type") +
  lgd + scale_fill_manual(values = c("Up" = "#F44336", "Down" = "#1976D2")) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1), 
        text = element_text(size = 22.5), 
        panel.border = element_rect(linewidth = 1)) +
  facet_wrap(. ~ Treatment, nrow = 1)
p
ggsave(plot = p, filename = paste0(opt$plot, "n_deg_per_celltype.png"), 
       height = 15, width = 50, units = "cm")

```

#### Show volcano plots 
```{r fig.height= 5.5, fig.width= 5.5}
geneSelec <- path.info %>% filter(pathway %in% c("NFKB", "NFKB_Chemo", "NFKB_GF", "NFKB_TF", "NFKB_Apo", "TNF"#, "TLR"
                                                 )) %>%
  select(gene) %>% unlist() %>% as.vector()

lapply(c("MCL", "HCL", "T-PLL", "Healthy T-Cells"), function(y) {
  pList <- lapply(c("Birinapant", "Everolimus", "Ibrutinib")[1], function(x) {

  volTab <- de.res.2020 %>% filter(Treatment == x, celltype == y) %>% #%>% dplyr::rename(Symbol = gene)
    mutate(p.adj = ifelse(p.adj < 1e-12, 1e-12, p.adj)) %>% arrange(p.adj)
  
  ## Define limits
  minLim <- -1 * max(abs(de.res.2020$logFC))
  maxLim <- max(abs(de.res.2020$logFC))
  thresh <- 0.25

  volTab$diffCol <- NA
  volTab[volTab$logFC > 0 & volTab$p.adj < 0.1 & volTab$logFC > thresh, "diffCol"] <- "sigUp"
  volTab[volTab$logFC < 0 & volTab$p.adj < 0.1 & volTab$logFC < -thresh, "diffCol"] <- "sigDown"
  volTab[volTab$p.adj >= 0.1, "diffCol"] <- "ns"
  volTab[volTab$logFC < thresh & volTab$logFC > -(thresh), "diffCol"] <- "ns"

  p <- volTab %>%
    mutate(logFC = ifelse(logFC > maxLim, maxLim, logFC)) %>%
    mutate(logFC = ifelse(logFC < minLim, minLim, logFC)) %>%
    #filter(gene %in% geneSelec) %>%
    filter(Treatment == x) %>%
    ggplot(aes(x = logFC, y = -log10(p.adj), fill = diffCol)) +
    geom_point(shape = 21, size = 2) + ggtitle(paste0(y, " - ", x)) +
    xlab("Log2FC") + ylab("-Log10(p.adj)") +
    xlim(-5, 5) + ylim(0, 12.75) +
    geom_hline(yintercept = -log10(0.1), linetype = "dashed") +
    geom_vline(xintercept = -thresh, linetype = "dashed") + geom_vline(xintercept = thresh, linetype = "dashed") +
    geom_label_repel(data = filter(volTab, Symbol %in% c(geneSelec), Treatment %in% x, p.adj < 0.1,
                                 logFC > thresh | logFC < -(thresh)), aes(label = Symbol),
                    nudge_y = 0.6, size = 5, max.overlaps = getOption("ggrepel.max.overlaps", default = 30),
                     segment.linetype = 2, force = 20, alpha = 0.8, fill = "white") +
    scale_fill_manual(values = c("sigUp" = "#F44336", "sigDown" = "#1976D2", "ns" = "grey")) +
    theme_classic() +
    theme(plot.title = element_text(hjust = 0.5),
          legend.position = "none",
          text = element_text(size = 17.5),
          axis.text = element_text(size = 15),
          panel.background = element_rect(fill = "transparent",colour = NA),
          plot.background = element_rect(fill = "transparent",colour = NA),
          legend.background = element_rect(fill='transparent'),
          strip.background = element_blank())

   ggsave(plot = p, filename = paste0(opt$plot, "vol_", y, "_", x, ".png"),
          height = 15, width = 15, unit = "cm")
  p
  })
  grid.arrange(grobs = pList)
})

```

#### Load per patient
```{r}
patSub <- names(sce.pat)
sce.pat <- mclapply(patSub, mc.cores = ncores, function(x) {
  sce <- sce.pat[[x]]

  set.seed(100)
  sce <- runUMAP(sce, n_neighbors = 15, min_dist = 0.35, name = "UMAP", BPPARAM = mcparam)

})
names(sce.pat) <- patSub


patSub <- c("H371", "H431", "H279")

lapply(patSub, function(x) {
  sce <- sce.pat[[x]]
  
  set.seed(100)
  sce <- sce[,sample(1:ncol(sce), ncol(sce))]
  lapply(c("Birinapant"), function(y) {
    plotTab <- ggcells(sce)[[1]]
    
  if(x == "H371") {tlt <- "T-PLL1"
  } else if(x == "H431") {tlt <- "T-PLL2"
  } else if(x == "H279") {tlt <- "T-PLL3"} else {
    tlt <- x
  }

    p <- plotTab %>% 
      filter(Treatment %in% c("DMSO", y)) %>%
      ggplot(aes(x = UMAP.1, y = UMAP.2, fill = Treatment)) +
      geom_point(shape = 21, size = 3) +
      ggtitle(tlt) +
      lgd +
      theme_void() +
      theme(axis.text = element_blank(), 
            axis.ticks = element_blank(), 
            legend.position = "none", 
            plot.title = element_text(size = 20, hjust = 0.5)) +
      scale_fill_manual(values = c("DMSO" = "grey70", "Birinapant" = "#AD1457", "Nutlin-3a" = "#FFA000", 
                                   "Ibrutinib" = "#283593", "Everolimus" = "#43A047", 
                                   "Selumetinib" = "#F8BBD0"))
    
      ggsave(plot = p, filename = paste0(opt$plot, x, "_perpat_", y, "_umap.png"), 
         height = 12.5, width = 12.5, units = "cm")
      p
  })
})

lapply(c("NFKBIA"), function(y) {
  
  lapply(c("Birinapant"), function(x) {
    lapply(patSub, function(z) {
        sce <- sce.pat[[z]]
        rownames(sce) <- rowData(sce)$Symbol
        plotTab <- ggcells(sce, features = y)[[1]]
      
    plotTab <- plotTab %>% 
      filter(Treatment %in% c("DMSO", x)) %>%
      mutate(Treatment = factor(Treatment, levels = c("DMSO", x))) %>%
      pivot_longer(cols = y, names_to = "gene", values_to = "expr") 
  
    p <- plotTab %>%
        ggplot(aes(x = UMAP.1, y = UMAP.2, fill = expr)) +
        geom_point(shape = 21, size = 3) +
        lgd +
        theme_void() +
        theme(axis.text = element_blank(),
              axis.ticks = element_blank(),
              legend.position = "none", 
              strip.text = element_text(size = 20, hjust = 0.5)) +
        scale_fill_gradient(high = "red", low = "grey80",
                            limits=c(0, quantile(plotTab$expr, 0.99)), oob=squish) +
      facet_wrap(. ~ Treatment)

      ggsave(plot = p, filename = paste0(opt$plot, "perpat/", x, "_perpat_", y, "_", z, "_umap.png"),
         height = 12.5, width = 27.5, units = "cm")
      
      p
      
    })
  })
})

```

#### Perform differential gene expression testing - scRNA-Seq - without subsampling
```{r message=FALSE, warning=FALSE}
opt$subsample <- FALSE

compDrug <- c("Nutlin-3a", "Birinapant", "Ibrutinib", "Everolimus", "Selumetinib")

sce_sub <- sce.sc
plotTab <- ggcells(sce_sub)[[1]]

colData(sce_sub) <- colData(sce_sub) %>%
  data.frame() %>%
  left_join(types, by = "uniqueBarcode") %>%
  mutate(celltype = ifelse(celltype %in% c("malignant B-cell", "malignant T-cell"), Diagnosis_simple, celltype)) %>%
  mutate(celltype = ifelse(celltype %in% c("healthy T-cell"), "T-Cells", celltype)) %>%
  mutate(celltype = ifelse(celltype %in% c("NK-cell"), "NK-Cells", celltype)) %>%
  mutate(celltype = ifelse(celltype %in% c("monocyte"), "Monocytes", celltype)) %>%
  mutate(celltype = ifelse(patientID%in% c("H501") & celltype %in% c("MCL", "HCL", "T-PLL"), "dead cell", celltype)) %>%
  mutate(celltype = ifelse(patientID%in% c("H431", "H371", "H279") & celltype %in% c("MCL", "HCL", "T-LGL"), "dead cell", celltype)) %>%
  mutate(celltype = ifelse(patientID%in% c("H526", "H525") & celltype %in% c("MCL", "T-PLL", "T-LGL"), "dead cell", celltype)) %>%
  mutate(celltype = ifelse(patientID%in% c("H496", "H432", "P1029") & celltype %in% c("HCL", "T-PLL", "T-LGL"), "dead cell", celltype)) %>%
  DataFrame()


if(opt$subsample == TRUE) {
 cellSub <- c("HCL", "MCL", "T-PLL", "T-Cells") 
} else {
  cellSub <- c("HCL", "MCL", "T-PLL")
}

testTreat <- c("Birinapant", "Nutlin-3a", "Ibrutinib", "Everolimus", "Selumetinib")

countTab <- plotTab %>%
  left_join(types, by = "uniqueBarcode") %>%
  mutate(celltype = ifelse(celltype %in% c("malignant B-cell", "malignant T-cell"), Diagnosis_simple, celltype)) %>%
  mutate(celltype = ifelse(celltype %in% c("healthy T-cell"), "T-Cells", celltype)) %>%
  mutate(celltype = ifelse(celltype %in% c("NK-cell"), "NK-Cells", celltype)) %>%
  mutate(celltype = ifelse(celltype %in% c("monocyte"), "Monocytes", celltype)) %>%
  group_by(celltype, Treatment) %>%
  dplyr::count() %>%
  filter(celltype %in% cellSub, Treatment %in% testTreat) %>%
  arrange(n)

minCells <- countTab[1, ]$n
minCells

de.res.2020 <- lapply(testTreat, function(z) {
  message(paste0("Fitting model for ", z))
  lapply(cellSub, #mc.cores = ncores, 
         function(x) {
    message(paste0("Celltype ", x))
    sce.x <- sce_sub[, sce_sub$celltype == x]
    sce.x <- sce.x[, sce.x$Treatment %in% c("DMSO", testTreat)]
    message(ncol(sce.x), " cells")
    
    if(opt$subsample == TRUE) {
      message("Subsetting to ", minCells)
      
      sce.x <- do.call(cbind, lapply(unique(sce.x$Treatment), function(y) {
        sce.y <- sce.x[, sce.x$Treatment == y]
        
        if(ncol(sce.y) == minCells) {sce.y} else {
          set.seed(99)
          sce.y <- sce.y[,sample(1:ncol(sce.y), minCells)]
          sce.y
        }
        sce.y
      }))
    }
    
    summed <- summarizeAssayByGroup(sce.x, ids = colData(sce.x)[,c("Diagnosis_simple", "patientID", "Treatment")], 
                                    statistics = "sum", assay.type = "counts") # get raw counts and sum them up
    #  summed <- applySCE(sce.x, aggregateAcrossCells, ids=colData(sce.x)[,c("Diagnosis_simple", "patientID", "Treatment")])
    colData(summed)$Treatment <- factor(colData(summed)$Treatment, levels = c("DMSO", compDrug))
    y <- DGEList(assay(summed), samples = colData(summed))
    opt$subsample
    if(opt$subsample == TRUE) {
      discarded <- summed$ncells < 5
    } else {
      discarded <- summed$ncells < 30
    }
    y <- y[,!discarded]
    message(paste0("Discarding ", sum(discarded), " samples"))
    design <- model.matrix(~0 + factor(patientID) + Treatment, y$samples) # set coefficient (name or matrix column) - fig. 7 f1000

    keep <- filterByExpr(y, design = design)
    y <- y[keep,]
    summary(keep)
    message(paste0("Keeping ", sum(keep), " genes"))
    y <- calcNormFactors(y)
    # head(design)
    y <- estimateDisp(y, design) # gene-wise dispersion (variance on top of poisson distribution)
    #plotBCV(y)
    fit <- glmQLFit(y, design, robust=TRUE) # QL way of fitting the model
    #plotQLDisp(fit)
    res <- glmQLFTest(fit, coef = paste0("Treatment", z))
    # #res <- glmQLFTest(fit, coef = "TreatmentNutlin-3a")
    summary(decideTests(res))
    resTab <- res$table
    resTab$gene <- rownames(resTab)
    resTab$celltype <- x
    resTab <- resTab #%>%
      #mutate(p.adj = ifelse(p.adj < 1e-12, 1e-12, p.adj))
    resTab$Treatment <- z
    resTab %>% arrange(desc(logFC)) %>% mutate(p.adj = p.adjust(PValue, method = "BH"))
  }) %>% bind_rows()
}) %>% bind_rows()

de.res.2020 <- de.res.2020 %>%
  left_join(select(data.frame(rowData(sce.sc)), ID, Symbol), c("gene" = "ID"))

```


#### Heatmap with logFC per patient
```{r fig.height= 10, fig.width=7.5}
opt$subsample <- FALSE

diffMat.list <- lapply(c("Birinapant"), function(z) {
  message(paste0("Fitting model for ", z))
  sce.x <- sce_sub[, sce_sub$celltype %in% c("T-PLL", "T-LGL", "HCL", "MCL")]
  colData(sce.x) <- colData(sce.x) %>%
    data.frame() %>%
    mutate(celltype = ifelse(celltype %in% c("T-PLL", "T-LGL", "HCL", "MCL"), patientID, celltype)) %>%
    DataFrame()
  sce.x <- sce.x[, sce.x$Treatment %in% c("DMSO", z)]
  summed <- summarizeAssayByGroup(sce.x, ids = colData(sce.x)[,c("celltype", "Treatment")], 
                                  statistics = "sum", assay.type = "counts") # get raw counts and sum them up

    colData(summed)$Treatment <- factor(colData(summed)$Treatment, levels = c("DMSO", compDrug))
    y <- DGEList(assay(summed), samples = colData(summed))
    opt$subsample
    if(opt$subsample == TRUE) {
      discarded <- summed$ncells < 10
    } else {
      discarded <- summed$ncells < 30
    }
    y <- y[,!discarded]
    message(paste0("Discarding ", sum(discarded), " samples"))
    design <- model.matrix(~0 + factor(celltype) + Treatment, y$samples) # set coefficient (name or matrix column) - fig. 7 f1000

    keep <- filterByExpr(y, design = design)
    y <- y[keep,]
    summary(keep)
    message(paste0("Keeping ", sum(keep), " genes"))
    y <- calcNormFactors(y)
    count.mat <- cpm(y, log = TRUE)
    colnames(count.mat) <- paste0(y$samples$celltype, "_", y$samples$Treatment)
    count.mat
  
    
      ## Compute the logFC per patient
  diffMat <- lapply(unique(sce.x$celltype), function(x) {
    message(x)
    ## Subset to the samples of this patient
    matSub <- count.mat[, str_detect(string = colnames(count.mat), pattern = x)]
    
    ## Save gene names
    matSub <- matSub %>% data.frame() %>% rownames_to_column("gene")
    matSub
    ## Compute logFC
    if(ncol(matSub) == 3) {
      colnames(matSub)[str_detect(string = colnames(matSub), pattern = "DMSO")] <- "DMSO"
      colnames(matSub)[str_detect(string = colnames(matSub), pattern = gsub(x = x, pattern = "-", replacement = "."))] <- "Treatment"

      matSub <- matSub %>%
        mutate(logFC = Treatment - DMSO) %>%
        select(gene, logFC)
    } else {
      matSub <- data.frame(gene = matSub$gene, logFC = NA)
    }
    colnames(matSub)[str_detect(string = colnames(matSub), pattern = "logFC")] <- x
    matSub %>% column_to_rownames("gene")
  }) %>% bind_cols()
})
names(diffMat.list) <- c("Birinapant")

geneSelec <- path.info %>% filter(pathway %in% c("NFKB", "NFKB_Chemo", "NFKB_GF", "NFKB_TF", "NFKB_Apo", "TNF"#, "TLR"
                                                 )) %>% 
  select(gene) %>% unlist() %>% as.vector()


patLabel <- data.frame(patientID = c("H371", "H431", "H279", "H501", "H452", "H432", "H496", "H525", "H526", "T-Cells"), 
                       tumorType = c("T-PLL1", "T-PLL2", "T-PLL3", "T-LGL1", "MCL1", "MCL2", "MCL3", "HCL1", "HCL2", "T-Cells"))

pHeat <- lapply(c("Birinapant")[[1]], function(x) {
  plotMat <- diffMat.list[[x]]
  
  if(x == "Nutlin-3a") {
    geneSelec <- path.info %>% filter(pathway %in% c("TP53up")) %>% select(gene) %>% unlist() %>% as.vector()
  }
  
  plotMat <- plotMat %>% rownames_to_column("ID") %>%
    left_join(select(data.frame(rowData(sce.sc)), ID, Symbol), c("ID")) %>%
    dplyr::select(-ID) %>% filter(Symbol %in% geneSelec) %>% column_to_rownames("Symbol")
  
  geneSelec <- intersect(rownames(plotMat), geneSelec)
  
  geneSig <- de.res.2020 %>% filter(p.adj < 0.1, Treatment %in% x, celltype %in% c("T-PLL", "MCL", "HCL")) %>% 
    filter(logFC > 0.25 | logFC < -0.25) %>%
    pull(Symbol) %>% unique()
  
  geneSelec <- intersect(geneSig, geneSelec)
  plotMat <- t(plotMat[geneSelec, ])
  rownames(plotMat) <- patLabel[match(rownames(plotMat), patLabel$patientID), ]$tumorType
  
  pHeat <- Heatmap(t(plotMat), clustering_method_rows = "ward.D2", 
          clustering_method_columns = "ward.D2", 
          show_row_names = TRUE,
          cluster_rows = TRUE, border = TRUE, #column_names_gp = gpar(fontsize = 8),
          column_title = paste0(x, " - Log2FC Malignant Cells"), 
          rect_gp = gpar(col = "black", lwd = 0.1),
          cluster_columns = TRUE, #row_names_gp = gpar(fontsize = 7),
          col=colorRamp2(c(-2, 0, 2), colors = c("#1976D2", "white", "#F44336")), name = "Log2FC")
  
  png(file=paste0(opt$plot, x, "_logFC_heatmap.png"),
      height = 25, width = 12.5, res = 600, units = "cm", bg = "transparent")
  draw(pHeat, background = "transparent")
  dev.off()
  
  pHeat
  
})
pHeat

```

#### Output session info
```{r pressure, echo=FALSE}
sessionInfo()
```
