---
title: "Supplemental Figure 4 & 5"
author: "M. Pohly"
date: "`r Sys.Date()`"
output: html_document
---

## Load libraries
```{r warning=FALSE, message=FALSE}
library(tidyverse)
library(scran)
library(scater)
library(gprofiler2)
library(gridExtra)
library(ComplexHeatmap)
library(circlize)
library(edgeR)
library(readxl)
library(ggrepel)
library(tftargets)
library(network)
library(ggnet)
library(parallel)
library(BiocParallel)
library(ggsci)
library(airr)
library(numbat)
library(scales)

ncores <- 10
mcparam <- MulticoreParam(workers=ncores)
register(mcparam)

```

## Define variables
```{r}
opt <- list()
opt$pathinfo <- "misc/pathways_2.xlsx"
opt$scesc <- "data/submission/scRNA_complete.RDS"
opt$scescpat <- "data/submission/scRNA_per_pat_complete.RDS"
opt$scebulk <- "data/submission/RNA_complete.RDS"
opt$plot <- "plots/SFig4_5/"


## Set theme
lgd <-  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5), 
        text = element_text(size = 15),
        panel.background = element_rect(fill = "transparent",colour = NA),
        plot.background = element_rect(fill = "transparent",colour = NA), 
        legend.background = element_rect(fill='transparent',colour = NA), 
        strip.background = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

```


## Load data
```{r}
## SCE
sce <- readRDS(opt$scebulk)
sce

## scRNA-Seq data
sce.sc <- readRDS(opt$scesc)

## single-cell cell types
types <- sce.sc[["types"]]

## Load tcr data
immTab <- sce.sc[["immTab"]]

sce.sc <- sce.sc[["SCE"]]

## sce per pat
sce.pat <- readRDS(opt$scescpat)

## Load pathway annotation
path.info <- read_excel(opt$pathinfo)
```


## Load transcription factors
```{r}
ntop <- 1000
maxtargets <- 5000

## Get TF target numbers
motif.df <- lapply(names(Marbach2016), function(x) {
  tf.list <- Marbach2016[[x]]
  #tf.list <- TRRUST[[x]]
  
  data.frame(motif_name = x, n_targets = length(tf.list))
  
}) %>% bind_rows()

## Get TF targets
target.df <- lapply(names(Marbach2016), function(x) {
  tf.list <- Marbach2016[[x]]
  
  data.frame(motif_name = x, targets = tf.list)
  
}) %>% bind_rows()

```

## Analysis
#### Perform differential gene expression testing
```{r}
testTreat <- unique(sce$Treatment)

de.res <- mclapply("DMSO", mc.cores = ncores, 
                 function(z) {
                   
  message(paste0("Fitting model for ", z))
                   
  compDrug <- testTreat[testTreat != z] # instead of subtypes we use other treatments here
                   
  mclapply(compDrug, mc.cores = ncores, 
         function(x) {
    message(paste0("Drug ", x))
    summed <- sce#[hvgenes, ]
    colData(summed)$Treatment <- factor(colData(summed)$Treatment, levels = c(z, compDrug))
    y <- DGEList(counts(summed), samples = colData(summed))
    design <- model.matrix(~0 + factor(patientID) + Treatment, y$samples) # set coefficient (name or matrix column) - fig. 7 f1000
    keep <- filterByExpr(y, design = design)
    y <- y[keep,]
    summary(keep)
    message(paste0("Keeping ", sum(keep), " genes"))
    y <- calcNormFactors(y)
    y <- estimateDisp(y, design) # gene-wise dispersion (variance on top of poisson distribution)
    plotBCV(y)
    fit <- glmQLFit(y, design, robust = TRUE) # QL way of fitting the model
    plotQLDisp(fit)
    res <- glmQLFTest(fit, coef = paste0("Treatment", x))
    summary(decideTests(res))
    resTab <- res$table
    resTab$gene <- rownames(resTab)
    resTab <- resTab %>% mutate(p.adj = p.adjust(PValue, method = "BH")) #%>%
      #mutate(p.adj = ifelse(p.adj < 1e-12, 1e-12, p.adj))
    resTab$Treatment <- x
    resTab$Control <- z
    resTab %>% arrange(desc(logFC))
  }) %>% bind_rows()
}) %>% bind_rows()

de.res <- de.res %>%
  left_join(select(data.frame(rowData(sce)), ID, Symbol), c("gene" = "ID"))

#write.csv(de.res, opt$deres)

sigTab <- de.res %>%
  group_by(Treatment) %>%
  filter(p.adj < 0.1) %>%
  filter(logFC > 0.25 | logFC < -0.25) %>%
  mutate(dir = ifelse(logFC > 0, "Up", "Down")) %>%
  dplyr::count(dir) %>%
  dplyr::rename(Direction = dir) %>%
  ungroup()

drugOrder <- sigTab %>%
  group_by(Treatment) %>%
  summarise(allDEG = sum(n)) %>%
  arrange(desc(allDEG)) %>%
  pull(Treatment)
```

#### Construct a network of differentially expressed genes
```{r fig.height= 10, fig.width=10, message=FALSE, warning=FALSE}
# Assign nodes (treatments) and edges (5% FDR DEGs)
plotMat <- de.res %>% filter(p.adj < 0.1) %>%
  filter(logFC > 0.25 | logFC < -0.25)
allTreat <- unique(plotMat$Treatment)
target <- data.frame(Treatment = allTreat,
                     target = c("Autophagyi", "IAPi", "HDACi", "mTORi", "ITKi", "TLR8a", "MDM2i", 
                                "IMID", "JAKi", "XPOi", "BCL2i"))

opt$sub <- TRUE

if(opt$sub == TRUE) {
  plotMat <- de.res %>% filter(p.adj < 0.1, !Treatment %in% c("Pomalidomide", "Dacinostat")
                             ) %>%
  filter(logFC > 0.25 | logFC < -0.25) 
} else {
  plotMat <- de.res %>% filter(p.adj < 0.1) %>%
    filter(logFC > 0.25 | logFC < -0.25)
}


allTreat <- unique(plotMat$Treatment)

nodes <- data.frame(Treatment = allTreat)
edges <- plotMat %>% select(Treatment, Symbol) %>%
  mutate(Treatment = as.character(Treatment))

# Create network object
set.seed(100)
net <- network(edges, directed = TRUE)
netTab <- ggnet2(net)$data |> 
  mutate(Type = ifelse(label %in% allTreat, "Treatment", "Symbol"),
         Treatment = ifelse(Type == "Treatment", label, NA)) |> 
  select(-alpha,-color,-shape,-size) |> 
  mutate(x = as.numeric(x), y = as.numeric(y))
  
# Define edges based on direction and padj of DEGs
from <- edges %>%
  left_join(netTab, by = c("Treatment" = "label")) %>% dplyr::select(Treatment, x, y) %>%
  dplyr::rename(x_start = x, y_start = y) |> distinct()

to <- edges %>%
  left_join(netTab, by = c("Symbol" = "label")) %>% dplyr::select(Symbol,x,y) %>%
  dplyr::rename(x_end = x, y_end = y) |> distinct()

edges <- edges %>% left_join(from) %>% left_join(to) %>%
  left_join(plotMat %>% select(Symbol,Treatment,logFC,p.adj)) %>%
  distinct() %>%
  mutate(Direction = ifelse(logFC < 0,"Down","Up"),
         alpha = -log10(p.adj)
         #alpha = rescale(-log10(p.adj#squish(p.adj, c(1.812407e-209,1))))
         )

# Create auxiliary table for annotation

n_DEG <- plotMat %>% #filter(!Treatment %in% c("Pomalidomide", "BafilomycinA1", "Dacinostat")) %>%
  group_by(Treatment) %>%
  filter(p.adj < 0.1) %>%
  filter(logFC > 0.25 | logFC < -0.25) %>% 
  dplyr::count() %>% ungroup()

netTabAnno <- netTab |> 
  left_join(target) |> 
  left_join(n_DEG) |> 
  mutate(Treatment = factor(Treatment, levels = allTreat))

p <- ggplot()  +
  geom_curve(data = edges,
               aes(x = x_start, xend = x_end, y = y_start, yend = y_end, col = Direction, alpha = p.adj),
               linewidth = 0.1, curvature = 0.05) +
  geom_point(data = netTabAnno[netTabAnno$Type == "Symbol",],
             aes(x = x, y = y),
             size = 0.5, shape = 21, color = "grey60", fill = "grey95", stroke = 0.75, alpha = 0.6) +
  geom_point(data = netTabAnno[netTabAnno$Type == "Treatment",],
             aes(x = x, y = y, fill = Treatment, size = n), shape = 21, stroke = 0.75) +
  # scale_fill_manual(values = col_TX) +
  scale_size_continuous(name = "Number of DEGs",
                        breaks = seq(0, max(n_DEG$n), by = 1000), limits = c(0, max(n_DEG$n)), range = c(0.5,7.5)) +
  scale_color_manual(values = c("Up" = "#F44336", "Down" = "#1976D2")) +
  scale_alpha_continuous() +
  geom_label_repel(data = netTabAnno |> filter(Type == "Treatment"),
                  aes(x = x, y = y, label = Treatment, fill = Treatment, point.size = rescale(n, to = c(0.5,7.5))),
                  max.overlaps = Inf, vjust = "bottom", nudge_y = -0.001, fontface = "bold", alpha = 0.7, #direction = "y", 
                  show.legend = FALSE, size = 5)  +
  guides(color = guide_legend(ncol = 2, override.aes = list(linewidth=2.5)),
         size = guide_legend(ncol = 2),
         alpha = guide_legend(override.aes = list(linewidth=2.5)),
         fill = guide_legend(override.aes = list(size=3.5))) +
  xlab("") + ylab("") + ggtitle("Network of Differentially Expressed Genes") +
  lgd +
  theme(axis.ticks = element_blank(), 
        axis.text = element_blank(), 
        panel.border = element_rect(linewidth = 1, colour = NA, fill = "transparent")) +
  scale_fill_manual(values = c("Venetoclax" = "grey70", "Birinapant" = "#AD1457", "Nutlin-3a" = "#FFA000", 
                                "Ibrutinib" = "#283593", "Everolimus" = "#43A047", "Selinexor" = "#FFD45F",
                               "BafilomycinA1" = "#F8BBD0", "Motolimod" = "#AB47BC", "Dacinostat" = "#F44336", 
                                "Dacinostat" = "#64B5F6", "Pomalidomide" = "#80DEEA", "Ruxolitinib" = "#4DB6AC"))

p
if(opt$sub == TRUE) {
 ggsave(plot = p, filename = paste0(opt$plot, "deg_subnetwork.png"),
       height = 20, width = 25, units = "cm")
} else {
  ggsave(plot = p, filename = paste0(opt$plot, "deg_network.png"),
       height = 20, width = 25, units = "cm")
}


```

#### Volcano plots
```{r fig.height = 5.5, fig.width = 5.5}
pList <- lapply(c("Motolimod"), function(x) {
  
  if(x %in% c("Birinapant", "Motolimod", "Everolimus", "Ibrutinib", "BaflimoycinA1")) {
    geneSelec <- path.info %>% filter(pathway %in% c("NFKB", "NFKB_Chemo", "NFKB_GF", "NFKB_TF", "NFKB_Apo", "TNF", "TLR")) %>% select(gene) %>% unlist() %>% as.vector()
  } else if(x %in% c("Nutlin-3a", "Selinexor")) {
    geneSelec <- path.info %>% filter(pathway %in% c("TP53up")) %>% select(gene) %>% unlist() %>% as.vector()
    geneSelec <- c("TP53", geneSelec)

  } else if(x %in% c("Everolimus")) {
    geneSelec <- path.info %>% filter(pathway %in% c("NFKB", "NFKB_Chemo", "NFKB_GF", "NFKB_TF", "NFKB_Apo", "TNF", "TLR")) %>% select(gene) %>% unlist() %>% as.vector()
  }
  
  volTab <- de.res %>% filter(Treatment == x)

  ## Define limits  
  minLim <- -1 * max(abs(de.res$logFC))
  maxLim <- max(abs(de.res$logFC))
  thresh <- 0.25

  volTab$diffCol <- NA
  volTab[volTab$logFC > 0 & volTab$p.adj < 0.1 & volTab$logFC > thresh, "diffCol"] <- "sigUp"
  volTab[volTab$logFC < 0 & volTab$p.adj < 0.1 & volTab$logFC < -thresh, "diffCol"] <- "sigDown"
  volTab[volTab$p.adj >= 0.1, "diffCol"] <- "ns"
  volTab[volTab$logFC < thresh & volTab$logFC > -(thresh), "diffCol"] <- "ns"
  
  p <- volTab %>%
    mutate(logFC = ifelse(logFC > maxLim, maxLim, logFC)) %>%
    mutate(logFC = ifelse(logFC < minLim, minLim, logFC)) %>%
    #filter(gene %in% geneSelec) %>%
    filter(Treatment == x) %>%
    ggplot(aes(x = logFC, y = -log10(p.adj), fill = diffCol)) +
    geom_point(shape = 21, size = 2) + xlab("Log2FC") + ylab("-Log10(p.adj)") + ggtitle(paste0(x)) +
    #xlim(minLim, maxLim) +
    geom_hline(yintercept = -log10(0.1), linetype = "dashed") +
    geom_vline(xintercept = -thresh, linetype = "dashed") + geom_vline(xintercept = thresh, linetype = "dashed") +
    geom_label_repel(data = filter(volTab, Symbol %in% c(geneSelec), Treatment %in% x, p.adj < 0.1, 
                                 logFC > thresh | logFC < -(thresh)), aes(label = Symbol), fill = "white", 
                    nudge_y = 0.25, size = 4, max.overlaps = getOption("ggrepel.max.overlaps", default = 30), 
                     segment.linetype = 2, force = 20, alpha = 0.8) +
    scale_fill_manual(values = c("sigUp" = "#F44336", "sigDown" = "#1976D2", "ns" = "grey")) +
    theme_classic() +
    theme(plot.title = element_text(hjust = 0.5), 
          legend.position = "none", 
          text = element_text(size = 17.5), 
          axis.text = element_text(size = 15),
          panel.background = element_rect(fill = "transparent",colour = NA),
          plot.background = element_rect(fill = "transparent",colour = NA),
          legend.background = element_rect(fill='transparent'),
          strip.background = element_blank())
  
   ggsave(plot = p, filename = paste0(opt$plot, "vol_", x, ".png"),
          height = 15, width = 15, unit = "cm")
  p

})
pList 

#### Plot bafilomycin A1 vs Motolimod
p <- de.res %>% filter(Treatment %in% c("BafilomycinA1", "Motolimod")) %>%
  select(gene, Symbol, logFC, Treatment) %>%
  pivot_wider(names_from = "Treatment", values_from = logFC) %>%
  ggplot(aes(x = BafilomycinA1, y = Motolimod)) +
  xlab("Bafilomycin A1 (log2FC)") + ylab("Motolimod (log2FC)") +
  ggtitle("Motolimod vs Bafilomycin A1") +
  geom_smooth(method = "lm", colour = "black", fill = "grey80") +
  geom_point() +
    theme_classic() +
    theme(plot.title = element_text(hjust = 0.5), 
          legend.position = "none", 
          text = element_text(size = 17.5), 
          axis.text = element_text(size = 15),
          panel.background = element_rect(fill = "transparent",colour = NA),
          plot.background = element_rect(fill = "transparent",colour = NA),
          legend.background = element_rect(fill='transparent'),
          strip.background = element_blank())

p
ggsave(plot = p, filename = paste0(opt$plot, "moto_vs_bafilo.png"),
       height = 15, width = 15, units = "cm")

```

#### TP53 dot plot
```{r fig.height = 25, fig.width = 7.5}
geneSelec <- path.info %>% filter(pathway %in% c("TP53up")) %>% select(gene) %>% unlist() %>% as.vector()

geneSub <- lapply(unique(de.res$Treatment), function(x) {
  de.res %>% 
    filter(Treatment == x) %>%
    filter(p.adj < 0.1) %>%
    filter(logFC > 0.25 | logFC < -0.25) %>%
    pull(Symbol)
  
}) %>% unlist() %>% unique()

geneSub <- intersect(geneSelec, geneSub)

plotMat.wide <- de.res %>% 
  mutate(logFC = ifelse(p.adj >= 0.1, 0, logFC)) %>%
  dplyr::select(Symbol, Treatment, logFC) %>% 
  filter(Symbol %in% geneSub) %>%
  dplyr::rename(log2FC = logFC) %>%
  pivot_wider(names_from = "Treatment", values_from = "log2FC") %>%
  column_to_rownames("Symbol")

geneOrder <- rownames(plotMat.wide)[hclust(dist(plotMat.wide), method = "ward.D2")$order]


p <- de.res %>% filter(#Treatment %in% c("Birinapant", "Motolimod"),
  Symbol %in% geneSub) %>%
  #filter(Symbol == "MDM2") %>%
  filter(p.adj < 0.1) %>%
  mutate(pSign = -log10(p.adj) * sign(logFC),
         p.adj = ifelse(-log10(p.adj) > 5, 5, -log10(p.adj)), 
         Symbol = factor(Symbol, levels = geneOrder)) %>%
  dplyr::rename(log2FC = logFC) %>%
  ggplot(aes(y = Symbol, x = Treatment)) +
  geom_point(aes(size = p.adj, fill = log2FC), shape = 21) +
  xlab("") + ylab("") + ggtitle("Effect on TP53 Target Genes") +
  lgd +
  scale_fill_gradient2(high = "#F44336", low = "#1976D2", name = "log2FC",
                       limits=c(-2, 2), oob=squish) +
  scale_size_continuous(name = "-Log10(p.adj)",
                        breaks = seq(0, 5, by = 2.5),
                        limits = c(0, 5)
                        ) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        text = element_text(size = 12.5),
        panel.border = element_rect(linewidth = 1))
p

ggsave(plot = p, filename = paste0(opt$plot, "tp53_long.png"),
       height = 40, width = 12.5, units = "cm")

```


#### GSEA
```{r fig.height = 5, fig.width = 12.5}

go.df.merge <- mclapply(c("Birinapant", "Motolimod", "Selinexor", "BafilomycinA1", "Dacinostat", "Pomalidomide", "Ibrutinib", 
                        "Everolimus", "Venetoclax", "Ruxolitinib", "Nutlin-3a"), mc.cores = ncores, 
                      function(compTreat) {
    message(paste("Running gprofiler for", compTreat, sep = " "))
    sigUp <- de.res %>% filter(logFC > 0.25, Treatment == compTreat, 
                               #!gene %in% c(drugOther), 
                               p.adj <= 0.1) %>% arrange(desc(logFC))
    #sigUp <- de.res %>% filter(celltype == compType, logFC > 0.25, Treatment == compTreat, p.adj <= 0.05) %>% arrange(PValue)
    
    sigUp
    if(nrow(sigUp) > 1) {
      gostresUp <- gost(query = sigUp$gene,
                        organism = "hsapiens", ordered_query = TRUE,
                        multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
                        measure_underrepresentation = FALSE, evcodes = FALSE,
                        user_threshold = 0.1, correction_method = "fdr",
                        domain_scope = "annotated", custom_bg = NULL,
                        numeric_ns = "", sources = NULL, as_short_link = FALSE)
      gostresUp$result

      go.df.up <- gostresUp$result %>% arrange(p_value)
      go.df.up$rank <- 1:nrow(go.df.up)
      go.df.up$type <- "Up"
    } else {
      go.df.up <- data.frame(p_value = NA, rank = NA, type = "Up", source = NA, term_name = NA, 
                               intersection_size = 10)
    }

    sigDown <- de.res %>% filter(logFC < -(0.25), Treatment == compTreat,
                                 #!gene %in% c(drugOther),
                                  p.adj <= 0.1) %>% arrange(logFC)
    #sigDown <- de.res %>% filter(celltype == compType, logFC < -(0.25), Treatment == compTreat, p.adj <= 0.05) %>% arrange(PValue)
    if(nrow(sigDown) > 1) {
    gostresDown <- gost(query = sigDown$gene,
                        organism = "hsapiens", ordered_query = TRUE,
                        multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
                        measure_underrepresentation = FALSE, evcodes = FALSE,
                        user_threshold = 0.1, correction_method = "fdr",
                        domain_scope = "annotated", custom_bg = NULL,
                        numeric_ns = "", sources = NULL, as_short_link = FALSE)

    gostresDown$result

    go.df.down <- gostresDown$result %>% arrange(p_value)
    go.df.down$rank <- 1:nrow(go.df.down) * -1
    go.df.down$type <- "Down"
    } else {
      go.df.down <- data.frame(p_value = NA, rank = NA, type = "Down", source = NA, term_name = NA, 
                               intersection_size = 10)
    }

    go.df.merge <- rbind(go.df.up, go.df.down)
    go.df.merge$Treatment <- compTreat

    go.df.merge <- go.df.merge %>% filter(!intersection_size < 10) %>% 
      filter(!source %in% c("CORUM", "TF", "MIRNA")) %>%
      mutate(p.adj = p.adjust(p_value, method = "BH"))
    
    if(nrow(go.df.merge) < 1) {
      go.df.merge <- data.frame(p_value = NA, rank = NA, type = "Down", source = NA, term_name = NA, 
                               intersection_size = 0)
    } else {
      go.df.merge
    }
}) %>% bind_rows() 

## Exclude all terms that are up and downregulated by a single-drug
dblTerm <- go.df.merge %>% 
  filter(!intersection_size < 10) %>%
  filter(!source %in% c("CORUM", "TF", "MIRNA")) %>%
  mutate(term_name = paste0(source, "_", term_name)) %>% unique() %>%
  #filter(term_name %in% unique(termTab$term_name)) %>%
  dplyr::count(term_name, Treatment) %>%
  filter(n > 1) %>%
  pull(term_name) %>% unique()

plotMat <- go.df.merge %>% 
  mutate(term_name = paste0(source, "_", term_name)) %>% unique() %>%
  filter(!source %in% c("CORUM", "TF", "MIRNA"), !intersection_size < 10) %>%
  filter(!term_name %in% c(dblTerm)) %>%
  #filter(term_name %in% gseaSub) %>%
  select(Treatment, term_name, type, p.adj) %>%
  mutate(p.adj = -log10(p.adj), 
         p.adj = ifelse(type == "Down", p.adj * -1, p.adj)) %>%
  select(-type) %>%
  pivot_wider(names_from = "term_name", values_from = "p.adj") %>%
  column_to_rownames("Treatment")

plotMat[is.na(plotMat)] <- 0

pHeat <- Heatmap(plotMat, show_column_names = FALSE, 
        clustering_method_rows = "ward.D2", 
        clustering_method_columns = "ward.D2", border = TRUE, 
        column_title = paste0("Enriched Gene Sets Among DEGs (FDR < 10%)"), 
        column_title_gp = gpar(fontsize = 17.5),
        heatmap_legend_param = list(title_gp = gpar(fontsize = 12.5)),
        col=colorRamp2(c(-5, 0, 5), colors = c("#1976D2", "white", "#F44336")), 
        name = "-Log10(p.adj)\nwith Direction")
pHeat
png(file=paste0(opt$plot, "GSEA_heatmap.png"),
    height = 10, width = 25, res = 600, units = "cm", bg = "transparent")
draw(pHeat, background = "transparent")
dev.off()

```

```{r fig.height = 7.5, fig.width = 7.5, message=FALSE, warning=FALSE}
## Plot top enriched pathwaysgo.df.merge
nTop <- 20

lapply(c("Selinexor"), function(y) {
           
  gseaTab <- go.df.merge %>% filter(Treatment == y) %>% filter(!source %in% c("CORUM", "TF", "MIRNA")) %>%
        mutate(term_name = paste0(source, "_", term_name)) %>% unique()
  
  message(y)
     lapply(unique(gseaTab$type), function(z) {
      message(z)
      
      gseaTab <- gseaTab %>% filter(type == z)
      
      gseaTab <- gseaTab[1:nTop, ] # get only top enrichments
      
      termOrder <- gseaTab %>% arrange(desc(p_value)) %>% pull(term_name)
      
      gseaTab <- gseaTab %>%
        mutate(term_name = factor(term_name, levels = termOrder), 
               p.adj = p.adjust(p_value, method = "BH"))
      
    if(-log10(min(gseaTab$p.adj, na.rm = TRUE)) > 10) {
      x.max <- -log10(min(gseaTab$p.adj, na.rm = TRUE))
    } else {
      x.max <- 10
    }
    if(y %in% c("Birinapant", "Selinexor", "Pomalidomide")) {
      x.max <- 3
    }
    if(y %in% c("Pomalidomide")) {
      x.max <- 2
    }


      
      p <- gseaTab %>%
        ggplot(aes(x = -log10(p.adj), y = term_name)) +
        geom_bar(stat = "identity", colour = "black", alpha = 0.5, aes(fill = type)) +
        geom_vline(xintercept = -log10(0.1), alpha = 0.8, linetype = "dashed") +
        geom_text(aes(label = term_name, y = term_name, x = 0.25), hjust = 0, size = 4.25) +
        xlab("-Log10(p.adj)") + ylab("") + #ggtitle(paste0("Top enriched gene-set: ", z)) +
        ggtitle(paste0(y, " - ", z)) +
        xlim(0, x.max) +
        lgd + theme(axis.text.y = element_blank(), 
                    axis.ticks.y = element_blank(), 
                    text = element_text(size = 22.5),
                    panel.border = element_rect(colour = "black", fill=NA, size=1.5), 
                    strip.background = element_blank(),
                    panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                    panel.background = element_rect(fill = "transparent",colour = NA),
                    plot.background = element_rect(fill = "transparent",colour = NA),
                    legend.background = element_rect(fill='transparent'), 
                    legend.position = "none") +
        scale_fill_manual(values = c("Down" = "#1976D2", "Up" = "#F44336"))
        #scale_fill_npg() +
  #guides(fill=guide_legend(title="Source"))
      ggsave(plot = p,
             file = paste0(opt$plot, "GSEA/", y, "_", z, "_top_", nTop, ".png"),
             height = 22.5, width = 19, units = "cm")
      p
    })
})

```


#### Hypergeometric test for transcription factors
```{r message=FALSE, warning=FALSE}
treatSub <- c("Pomalidomide")

## Run hypergeometric test
hyper.df <- lapply(treatSub, function(x) {
  ## Limit the number of targets
  motif.df <- motif.df %>% filter(n_targets < maxtargets)
  message("Keeping ", nrow(motif.df), " TFs with less than ", maxtargets)

  ## Get the transcription factor targets
  lapply(c("Up", "Down"), function(y) {
    df.hub <- de.res %>% filter(p.adj < 0.05, Treatment == x)
    
    if(y == "Up") {df.hub <- df.hub %>% filter(logFC > 0.25)}
    if(y == "Down") {df.hub <- df.hub %>% filter(logFC < -0.25)}
    
    mclapply(unique(motif.df$motif_name), mc.cores = ncores, 
             function(z) {
      message(z)
      tf.targets <- target.df %>% filter(motif_name == z) %>% pull(targets)

      size.module <- nrow(df.hub)
      size.genome <- length(unique(de.res$Symbol))
      size.genome <- nrow(sce)
      n.targets <- length(tf.targets)
      n.over <- intersect(df.hub$Symbol, tf.targets) %>% length()

      p.val <- phyper(q = n.over - 1, # number of white balls drawn
                      m = n.targets, # number white balls in urn
                      n = size.genome - n.targets, # number black balls in urn
                      k = size.module, # number of draws
                      lower.tail= FALSE)

      df <- data.frame(Treatment = x, dir = y, tf = z, p.val = p.val, n.over = n.over,
                       n.targets = n.targets, size.module = size.module,
                       size.genome = size.genome)
      #df %>% filter(n.over != 0)
    }) %>% bind_rows()
  }) %>% bind_rows() %>% mutate(p.adj = p.adjust(p.val, method = "BH"))
})# %>% bind_rows()

names(hyper.df) <- treatSub

lapply(c("Pomalidomide"), function(x) {
  lapply(c("Up", "Down"), function(y) {
    message(x)
    hyper.df <- hyper.df %>% bind_rows() %>%
      filter(Treatment == x, dir == y) %>%
      arrange(p.adj) %>%
      filter(p.adj < 0.1) %>%
      mutate(p.adj = -log10(p.adj))

    hyper.df$rank <- 1:nrow(hyper.df)

    ## Get the top 5 enriched TPs
    toptf <- hyper.df %>% filter(rank <= 10) %>%
      pull(tf)

    if(x == "Pomalidomide") {
      toptf <- c(toptf, "IKZF1", "IKFZ3", "BRD7", "ARID2", "IRF4", "MYC")
    }

    if(x %in% c("Nutlin-3a", "Selinexor")) {
      toptf <- c(toptf, "TP53")
    }
    if(x %in% c("Birinapant", "Motolimod", "BafilomycinA1", "Ibrutinib")) {
      toptf <- c(toptf, "NFKB", "NFKB1", "NFKB2", "REL", "RELA", "RELB")
    }

    if(y == "Up") {col.point <- "#F44336"}
    if(y == "Down") {col.point <- "#1976D2"}

    p <- hyper.df %>%
      ggplot(aes(x = rank, y = p.adj)) +
      geom_point() +
      geom_point(data = hyper.df[hyper.df$tf %in% toptf, ], fill = col.point, size = 2.5, shape = 21) +
      xlab("Rank") + ylab("-Log10(p.adj)") + ggtitle(paste0(x, " - TF Enrichment (", y, ")")) +
      geom_label_repel(aes(label = tf), data = hyper.df[hyper.df$tf %in% toptf, ],
                       max.overlaps = getOption("ggrepel.max.overlaps", default = 30),
                       segment.linetype = 2, force = 20, alpha = 0.8) +
      lgd +
      theme_classic() +
      theme(plot.title = element_text(hjust = 0.5),
            legend.position = "none",
            text = element_text(size = 17.5),
            axis.text = element_text(size = 15),
            panel.background = element_rect(fill = "transparent",colour = NA),
            plot.background = element_rect(fill = "transparent",colour = NA),
            legend.background = element_rect(fill='transparent'),
            strip.background = element_blank())

    ggsave(plot = p, filename = paste0(opt$plot, x, "_", y, "_tf_enrichment.png"),
           height = 12.5, width = 16, units = "cm")

    p
  })
})

```


#### Overview scRNA-Seq data
```{r}
## Randomly sample 
set.seed(100)
sce.sc <- sce.sc[, sample(1:ncol(sce.sc), ncol(sce.sc))]

## Re-run UMAP
set.seed(100)
sce.sc <- runUMAP(sce.sc, n_neighbors = 15, min_dist = 0.75, name = "UMAP", BPPARAM = mcparam)

## Make plots
plotTab <- ggcells(sce.sc)[[1]]

## Pairwise each treatment
lapply(c("Birinapant", "Nutlin-3a", "Ibrutinib", "Everolimus", "Selumetinib"), function(x) {
  p <- plotTab %>%
    filter(Treatment %in% c("DMSO", x)) %>%
    ggplot(aes(x = UMAP.1, y = UMAP.2, fill = Treatment)) +
    geom_point(shape = 21, size = 2) +
    lgd + theme_void() + theme(legend.position = "none") + 
    scale_fill_manual(values = c("DMSO" = "grey70", "Birinapant" = "#AD1457", "Nutlin-3a" = "#FFA000", 
                                 "Ibrutinib" = "#283593", "Everolimus" = "#43A047", 
                                 "Selumetinib" = "#F8BBD0"))
  
    ggsave(plot = p, filename = paste0(opt$plot, x, "_scrna_umap.png"), 
         height = 15, width = 15, units = "cm")
    p
})

```

#### Show donut plots
```{r}

plotTab.vdj <- plotTab %>%
  left_join(types, by = "uniqueBarcode") %>%
  filter(celltype %in% c("malignant B-cell", "healthy B-cell")) %>%
  mutate(celltype = ifelse(celltype == "malignant B-cell", patientID, celltype), 
         celltype = ifelse(celltype == "healthy B-cell", "Healthy B-Cells", celltype))

plotTab.vdj[plotTab.vdj$celltype == "H525", "celltype"] <- "HCL1"
plotTab.vdj[plotTab.vdj$celltype == "H526", "celltype"] <- "HCL2"
plotTab.vdj[plotTab.vdj$celltype == "H432", "celltype"] <- "MCL1"
plotTab.vdj[plotTab.vdj$celltype == "H452", "celltype"] <- "MCL2"
plotTab.vdj[plotTab.vdj$celltype == "H496", "celltype"] <- "MCL3"

## Merge VDJ data with plotting info
plotTab.vdj <- plotTab.vdj %>% left_join(select(immTab, cell_id, clone_id, c_call#, mu_freq
                                                ), 
                                     by = c("uniqueBarcode" = "cell_id"))

plotTab.vdj %>% filter(clone_id %in% c("60_5")) %>% select(celltype)

## Calculate the number of clones per cell type
countTab <- plotTab.vdj %>%
  group_by(celltype) %>%
  select(celltype, clone_id) %>% unique() %>%
  dplyr::count(celltype) %>% mutate(n_clones = n) %>%
  select(-n)
countTab

## Calculate the size of each clone
countSize <- plotTab.vdj %>%
  filter(!is.na(clone_id)) %>%
  group_by(celltype) %>%
  select(celltype, clone_id) %>%
  dplyr::count(celltype, clone_id) %>% mutate(clone_size = n) %>%
  select(-n)
countSize

###################
# Donut charts of clonal composition
###################
## Calculate the size of each clone - don't subset to only cells with clone_id
# otherwise NA T-cells will not be in the plot later.
cloneCount <- plotTab.vdj %>% 
  group_by(celltype) %>%
  dplyr::count(clone_id) %>% ungroup()
cloneCount

cloneNum <- lapply(unique(plotTab.vdj$celltype), function(x) {
  subTab <- filter(plotTab.vdj, celltype == x) # subset to each patient
  countTab <- dplyr::count(subTab, clone_id)
  nclones <- countTab %>% nrow() # get the number of clones
  ncells_bcr <- nrow(subTab) # calculate the number of B or T cells per patient - important to subset plotTab to only B or T-cell clusters before doing this
  res <- data.frame(patID = x, nclones = nclones, ncells_bcr = ncells_bcr)
  res$type <- opt$subType
  res
}) %>% bind_rows()
cloneNum

df <- left_join(cloneCount, cloneNum, by = c("celltype" = "patID")) %>%
  group_by(celltype) %>% mutate(clone_ratio = n / ncells_bcr) %>% # divide the size of the clone 
  select(celltype, clone_id, clone_ratio, nclones) %>% unique() %>%
  mutate(ymax = cumsum(clone_ratio), ymin = c(0, head(ymax, n= -1)))
df

#### Overview over all patients/celltypes
pClone.Pat <- df %>% 
  ggplot(aes(ymax = ymax, ymin = ymin, xmax=4, xmin=3, fill = clone_id)) +
  geom_rect(colour = "black", size = 0.25) +
  coord_polar(theta="y") +
  xlim(c(2, 4)) +
  # ggtitle(paste0("Clonal Composition per Patient - ", opt$subType)) +
  facet_wrap(. ~ celltype, nrow =  1) +
  theme_void() +
  theme(plot.title = element_text(hjust = 0.5),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 12.5),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_rect(fill = "transparent",colour = NA),
        plot.background = element_rect(fill = "transparent",colour = NA),
        legend.position = "none")

pClone.Pat

#### Output individual patients
subPat <- c("HCL1", "HCL2", "MCL1", "MCL2", "MCL3", "Healthy B-Cells")

lapply(subPat, function(subPat) {
  pClone <- df %>% filter(celltype == subPat) %>%
  ggplot(aes(ymax = ymax, ymin = ymin, xmax=4, xmin=3, fill = clone_id)) +
  geom_rect(colour = "black", size = 0.25) +
  coord_polar(theta="y") +
  xlim(c(2, 4)) +
  ggtitle(paste0(subPat)) +
  theme_void() +
  theme(plot.title = element_text(hjust = 0.5, size = 75),
        legend.position = "none")
  pClone
  
  ggsave(plot = pClone, filename = paste0(opt$plot, "bcr_comp_", subPat, ".png"), 
         height = 20, width = 20, units = "cm")
})

```

#### Per patient UMAPs
```{r}
patSub <- names(sce.pat)
sce.pat <- mclapply(patSub, mc.cores = ncores, function(x) {
  sce <- sce.pat[[x]]

  set.seed(100)
  sce <- runUMAP(sce, n_neighbors = 15, min_dist = 0.35, name = "UMAP", BPPARAM = mcparam)

})
names(sce.pat) <- patSub

patSub <- c("H371", "H431", "H279")

lapply(patSub, function(x) {
  sce <- sce.pat[[x]]
  
  set.seed(100)
  sce <- sce[,sample(1:ncol(sce), ncol(sce))]
  plotTab <- ggcells(sce)[[1]]
  
  if(x == "H371") {tlt <- "T-PLL1"
  } else if(x == "H431") {tlt <- "T-PLL2"
  } else if(x == "H279") {tlt <- "T-PLL3"} else {
    tlt <- x
  }  
  
  p <- plotTab %>% 
    ggplot(aes(x = UMAP.1, y = UMAP.2, fill = Treatment)) +
    geom_point(shape = 21, size = 3) +
    ggtitle(tlt) +
    lgd +
    theme_void() +
    theme(axis.text = element_blank(), 
          axis.ticks = element_blank(), 
          legend.position = "none", 
          plot.title = element_text(size = 20, hjust = 0.5)) +
    scale_fill_manual(values = c("DMSO" = "grey70", "Birinapant" = "#AD1457", "Nutlin-3a" = "#FFA000", 
                                 "Ibrutinib" = "#283593", "Everolimus" = "#43A047", 
                                 "Selumetinib" = "#F8BBD0"))
  
  ggsave(plot = p, filename = paste0(opt$plot, x, "_perpat_umap.png"), 
         height = 12.5, width = 12.5, units = "cm")
  
  p
})

```

#### Output session info
```{r pressure, echo=FALSE}
sessionInfo()
```
